import{_ as e,c as a,o as n,a3 as t}from"./chunks/framework.BXSIK7tN.js";const m=JSON.parse('{"title":"Pipeline Information","description":"","frontmatter":{},"headers":[],"relativePath":"pipeline_information.md","filePath":"pipeline_information.md"}'),s={name:"pipeline_information.md"},i=t(`<h1 id="pipeline-information" tabindex="-1">Pipeline Information <a class="header-anchor" href="#pipeline-information" aria-label="Permalink to &quot;Pipeline Information&quot;">​</a></h1><p>To link barcodes to specific variants present in each RBP sequence, we did Pacbio sequencing on RBP variant libraries. Full-length RBP consensus sequences were made to determine which mutations were present in each library. DMS data was analyzed with <a href="https://github.com/dms-vep/dms-vep-pipeline-3" target="_blank" rel="noreferrer">dms-vep-pipeline-3</a>. This pipeline utilizes the <a href="https://jbloomlab.github.io/alignparse/" target="_blank" rel="noreferrer">alignparse</a> python package. The general steps are listed below.</p><h2 id="build-pacbio-sequences" tabindex="-1">Build Pacbio Sequences <a class="header-anchor" href="#build-pacbio-sequences" aria-label="Permalink to &quot;Build Pacbio Sequences&quot;">​</a></h2><p><a href="notebooks/build_pacbio_consensus.html" target="_self">Build PacBio consensus sequences notebook</a></p><p>Builds the pacbio consensus sequence. Paramaters used are:</p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code"><code><span class="line"><span>max_minor_sub_frac=0.2</span></span>
<span class="line"><span>max_minor_indel_frac=0.2</span></span>
<span class="line"><span>min_support=3</span></span></code></pre></div><p>These parameters filter consensus sequences made from pacbio CCS sequencing. If a consensus RBP sequence has a mutation or indel at greater than 20% frequency of the reads, it will be discarded. Consensus sequences must have at least three reads to be included.</p><h2 id="analyze-pacbio-ccs-reads" tabindex="-1">Analyze PacBio CCS Reads <a class="header-anchor" href="#analyze-pacbio-ccs-reads" aria-label="Permalink to &quot;Analyze PacBio CCS Reads&quot;">​</a></h2><p><a href="notebooks/analyze_pacbio_ccs.html" target="_self">Analyze Pacbio CCS reads notebook</a></p><p>Reports stats on how many CCS reads were filtered out or aligned to the unmutated reference sequence.</p><h2 id="build-codon-variants-notebook" tabindex="-1">Build Codon Variants Notebook <a class="header-anchor" href="#build-codon-variants-notebook" aria-label="Permalink to &quot;Build Codon Variants Notebook&quot;">​</a></h2><p><a href="notebooks/build_codon_variants.html" target="_self">Build codon variants notebook</a></p><p>Builds the codon-variant table that links barcodes and mutations.</p><h2 id="illumina-variant-counts" tabindex="-1">Illumina Variant Counts <a class="header-anchor" href="#illumina-variant-counts" aria-label="Permalink to &quot;Illumina Variant Counts&quot;">​</a></h2><p>Once the barcodes are linked to mutations in the codon-variant table, all sequencing data is generated with Illumina to get the relative frequencies of barcodes in each experiment.</p><p><a href="notebooks/analyze_variant_counts.html" target="_self">Analysis of variant counts notebook</a></p><h2 id="filtering-selection-data" tabindex="-1">Filtering Selection Data <a class="header-anchor" href="#filtering-selection-data" aria-label="Permalink to &quot;Filtering Selection Data&quot;">​</a></h2><p><a href="notebooks/filter_data.html" target="_self">Filtering notebook</a></p><p>The final averaged data were filtered based on the following parameters that are contained within the <a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/nipah_config.yaml" target="_blank" rel="noreferrer">nipah_config.yaml</a> file. More information about these parameters are listed below.</p><p><strong>Cell Entry</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code"><code><span class="line"><span>func_times_seen_cutoff=2 A mutation must be observed in at least two separate barcodes (the average times_seen in the averaged effects) to be included.</span></span>
<span class="line"><span>func_std_cutoff=1 The standard deviation of the effect of the mutation must be below 1.</span></span>
<span class="line"><span>percent_selections=0.5 A mutation must be present in at least 50% of the individual selections to be included.</span></span></code></pre></div><p><strong>Receptor Binding</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code"><code><span class="line"><span>min_times_seen_binding=2.5 A mutation must be observed in at least 2.5 separate barcodes to be included.</span></span>
<span class="line"><span>max_binding_std=2 The standard deviation of the effect of the mutation must be below 2.</span></span>
<span class="line"><span>min_func_effect_for_binding=-1.5 A mutation must have a cell entry score of greater than -1.5 to be included in the receptor binding analyses.</span></span></code></pre></div><p><strong>Antibody Escape</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code"><code><span class="line"><span>min_times_seen_ab=2.5 A mutation must be observed in at least 2.5 separate barcodes to be included.</span></span>
<span class="line"><span>max_ab_std=2 The standard deviation of the effect of the mutation must be below 2.</span></span>
<span class="line"><span>frac_models=0.5 A mutation must be present in at least 50% of the individual selections</span></span>
<span class="line"><span>min_func_effect_for_ab=-2 A mutation must have a cell entry score of at least -2 to be included in the antibody escape analyses.</span></span></code></pre></div><h2 id="miscellaneous-notebooks" tabindex="-1">Miscellaneous Notebooks <a class="header-anchor" href="#miscellaneous-notebooks" aria-label="Permalink to &quot;Miscellaneous Notebooks&quot;">​</a></h2><p><a href="notebooks/make_nipah_phylogeny_baltic.html" target="_self">Notebook for making a Nipah phylogeny</a></p><p><a href="notebooks/mapping_site_level.html" target="_self">Notebook for mapping filtered scores onto protein structures</a></p><p><a href="notebooks/receptor_distance.html" target="_self">Calculate atomic distances between residues in a PDB file notebook</a></p><p><a href="notebooks/henipavirus_conservation.html" target="_self">Find variable sites in Nipah or Henipavirus alignments notebook</a></p>`,30),o=[i];function l(r,p,c,d,h,u){return n(),a("div",null,o)}const f=e(s,[["render",l]]);export{m as __pageData,f as default};
