{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "262ca3d4-3cc9-4175-be21-210319a490a6",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### This script analyzes filtered mAb escape data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b735289-ae49-4ed1-b375-99eb52b63fbf",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as parameters for `papermill` parameterization\n",
    "binding_data = None\n",
    "HENV103_filter = None\n",
    "HENV117_filter = None\n",
    "HENV26_filter = None\n",
    "HENV32_filter = None\n",
    "m102_filter = None\n",
    "nAH1_filter = None\n",
    "\n",
    "altair_config = None\n",
    "nipah_config = None\n",
    "escape_bubble_plot = None\n",
    "bubble_1_mut_plot = None\n",
    "mab_line_escape_plot = None\n",
    "aggregate_mab_and_binding = None\n",
    "aggregate_mab_and_niv_polymorphism = None\n",
    "binding_vs_escape = None\n",
    "\n",
    "mab_plot_top = None\n",
    "mab_plot_all = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "166966a9-1614-49cd-9c0b-e36cccd2b287",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "if binding_data is None:\n",
    "    print(\"this is being run manually\")\n",
    "else:\n",
    "    print(\"papermill!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f83e70ff-0049-4b6b-a852-bc4d9034333b",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import math\n",
    "import os\n",
    "import re\n",
    "\n",
    "import altair as alt\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import scipy.stats\n",
    "\n",
    "import Bio.SeqIO\n",
    "import yaml\n",
    "import matplotlib\n",
    "\n",
    "matplotlib.rcParams[\"svg.fonttype\"] = \"none\"\n",
    "\n",
    "from Bio import PDB\n",
    "import dmslogo\n",
    "from dmslogo.colorschemes import CBPALETTE\n",
    "from dmslogo.colorschemes import ValueToColorMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "402f8b06-af68-47b0-90ce-d3aeef00a455",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# allow more rows for Altair\n",
    "_ = alt.data_transformers.disable_max_rows()\n",
    "\n",
    "if (\n",
    "    os.getcwd()\n",
    "    == \"/fh/fast/bloom_j/computational_notebooks/blarsen/2023/Nipah_Malaysia_RBP_DMS/\"\n",
    "):\n",
    "    pass\n",
    "    print(\"Already in correct directory\")\n",
    "else:\n",
    "    os.chdir(\n",
    "        \"/fh/fast/bloom_j/computational_notebooks/blarsen/2023/Nipah_Malaysia_RBP_DMS/\"\n",
    "    )\n",
    "    print(\"Setup in correct directory\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45063f53-6abc-444e-b1c1-4cacd61da1ef",
   "metadata": {},
   "source": [
    "### For running interactively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1dfb73e-adfa-4405-9f1c-4612adf8703a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "if binding_vs_escape is None:\n",
    "    altair_config = \"data/custom_analyses_data/theme.py\"\n",
    "    nipah_config = \"nipah_config.yaml\"\n",
    "\n",
    "    binding_data = \"results/filtered_data/E2_binding_filtered.csv\"\n",
    "\n",
    "    HENV103_filter = \"results/filtered_data/HENV103_escape_filtered.csv\"\n",
    "    HENV117_filter = \"results/filtered_data/HENV117_escape_filtered.csv\"\n",
    "    HENV26_filter = \"results/filtered_data/HENV26_escape_filtered.csv\"\n",
    "    HENV32_filter = \"results/filtered_data/HENV32_escape_filtered.csv\"\n",
    "    m102_filter = \"results/filtered_data/m102_escape_filtered.csv\"\n",
    "    nAH1_filter = \"results/filtered_data/nAH1_escape_filtered.csv\"\n",
    "\n",
    "    # escape_bubble_plot = 'results/images/escape_bubble_plot.html'\n",
    "    # bubble_1_mut_plot = 'results/images/escape_bubble_1_mut_plot.html'\n",
    "    # overlap_escape_plot = 'results/images/overlap_escape_plot.html'\n",
    "\n",
    "    # m102_heat = 'results/images/m102_heatmap.html'\n",
    "    # HENV26_heat = 'results/images/HENV26_heatmap.html'\n",
    "    # HENV32_heat = 'results/images/HENV32_heatmap.html'\n",
    "    # nAH1_heat = 'results/images/nAH1_heatmap.html'\n",
    "    # HENV117_heat = 'results/images/HENV117_heatmap.html'\n",
    "    # HENV103_heat = 'results/images/HENV103_heatmap.html'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a592e22-21fe-4a6f-a472-2b241896207d",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "if altair_config:\n",
    "    with open(altair_config, \"r\") as file:\n",
    "        exec(file.read())\n",
    "\n",
    "with open(nipah_config) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edacfe37-0325-4426-817e-bf56e42e5579",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "# Make logo plots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79764084-978e-4640-b424-81350df7c729",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Filtering parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b418f18-bb9f-42b2-a86c-1ecb255deb0a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make a dataframe with all the mutants with low entry scores for masking later in script\n",
    "func_scores_E3 = pd.read_csv(\n",
    "    \"../Nipah_Malaysia_RBP_DMS/results/func_effects/averages/CHO_EFNB3_low_func_effects.csv\"\n",
    ")\n",
    "func_scores_E3_low_effect = func_scores_E3[\n",
    "    (func_scores_E3[\"effect\"] < config[\"min_func_effect_for_ab\"])\n",
    "    & (func_scores_E3[\"times_seen\"] > config[\"func_times_seen_cutoff\"])\n",
    "    & (func_scores_E3[\"site\"] != 603)\n",
    "    & (func_scores_E3[\"mutant\"] != \"-\")\n",
    "    & (func_scores_E3[\"mutant\"] != \"*\")\n",
    "]\n",
    "display(func_scores_E3_low_effect)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b31686c8-f108-4e24-9935-9477fccfaea6",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Read in filtered antibody escape files and combine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c180ce3f-afa9-4811-a3c4-1e9d44dcd63e",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "HENV103 = pd.read_csv(HENV103_filter)\n",
    "HENV117 = pd.read_csv(HENV117_filter)\n",
    "HENV26 = pd.read_csv(HENV26_filter)\n",
    "HENV32 = pd.read_csv(HENV32_filter)\n",
    "m102 = pd.read_csv(m102_filter)\n",
    "nAH1 = pd.read_csv(nAH1_filter)\n",
    "\n",
    "# Combine all the individual filtered antibody escape files\n",
    "combined_df = pd.concat([HENV103, HENV117, HENV26, HENV32, m102, nAH1])\n",
    "combined_df = combined_df[\n",
    "    [\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"mutant\",\n",
    "        \"mutation\",\n",
    "        \"effect\",\n",
    "        \"escape_median\",\n",
    "        \"escape_std\",\n",
    "        \"times_seen_ab\",\n",
    "        \"show_site\",\n",
    "        \"ab\",\n",
    "    ]\n",
    "]\n",
    "display(combined_df)\n",
    "\n",
    "# Make a separate dataframe that only has the top sites\n",
    "filtered_df = combined_df.query(\"show_site == True\")\n",
    "filtered_df = filtered_df[filtered_df[\"escape_median\"] >= config[\"min_escape_cutoff\"]]\n",
    "display(filtered_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb2a1da0-feef-402c-893c-e173d94a8f7f",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def identify_escape_sites(df, ab):\n",
    "    subset = df[(df[\"ab\"] == ab)]\n",
    "    unique_sites = list(subset[\"site\"].unique())\n",
    "    return unique_sites\n",
    "\n",
    "\n",
    "abs = [\"HENV-26\", \"HENV-103\", \"HENV-32\", \"HENV-117\", \"m102.4\", \"nAH1.3\"]\n",
    "sites_dict = {}  # Create an empty dictionary to store the results\n",
    "\n",
    "for ab in abs:\n",
    "    sites_dict[ab] = identify_escape_sites(filtered_df, ab)\n",
    "\n",
    "display(sites_dict)  # need site dict for later"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b2cffcf-e757-4cef-9083-41b43557a503",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Plot bubble chart showing mAb escape for individual mutants by functional score for both E2 or E3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6fdff17-3f32-4059-89d3-ae2d5c8f5542",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "order_ab = [\"m102.4\", \"HENV-26\", \"HENV-117\", \"HENV-103\", \"HENV-32\", \"nAH1.3\"]\n",
    "\n",
    "\n",
    "def generate_chart(df):\n",
    "    variant_selector = alt.selection_point(\n",
    "        on=\"mouseover\", empty=False, fields=[\"site\"], value=1\n",
    "    )\n",
    "    chart = (\n",
    "        alt.Chart(\n",
    "            df,\n",
    "            title=alt.Title(\n",
    "                \"Top Antibody Escape Mutations\",\n",
    "                subtitle=\"Hover over points to see escape at same site\",\n",
    "            ),\n",
    "        )\n",
    "        .mark_point(stroke=\"black\")\n",
    "        .encode(\n",
    "            x=alt.X(\n",
    "                \"ab:O\",\n",
    "                sort=order_ab,\n",
    "                title=\"Antibody\",\n",
    "                axis=alt.Axis(labelAngle=-90, grid=False),\n",
    "            ),\n",
    "            y=alt.Y(\n",
    "                \"effect:Q\",\n",
    "                title=\"Cell Entry of Top Escape\",\n",
    "                axis=alt.Axis(\n",
    "                    grid=True, tickCount=4, values=[0.5, 0, -0.5, -1, -1.5, -2]\n",
    "                ),\n",
    "            ),\n",
    "            size=alt.Size(\n",
    "                \"escape_median\", legend=alt.Legend(title=\"Mean Escape By Mutation\")\n",
    "            ),\n",
    "            xOffset=\"random:Q\",\n",
    "            tooltip=[\n",
    "                \"site\",\n",
    "                \"wildtype\",\n",
    "                \"mutant\",\n",
    "                \"ab\",\n",
    "                \"effect\",\n",
    "                \"escape_median\",\n",
    "                \"escape_std\",\n",
    "            ],\n",
    "            color=alt.Color(\"ab\").legend(None),\n",
    "            opacity=alt.condition(variant_selector, alt.value(1), alt.value(0.2)),\n",
    "            strokeWidth=alt.condition(variant_selector, alt.value(2), alt.value(0)),\n",
    "        )\n",
    "        .transform_calculate(\n",
    "            random=\"sqrt(-1*log(random()))*cos(2*PI*random())\"\n",
    "        )\n",
    "        .properties(width=config[\"bubble_width\"], height=config[\"bubble_height\"])\n",
    "        .add_params(variant_selector)\n",
    "    )\n",
    "\n",
    "    return chart\n",
    "\n",
    "\n",
    "escape_bubble = generate_chart(filtered_df)\n",
    "escape_bubble.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    escape_bubble.save(escape_bubble_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f8e4a8e-1ef3-46fa-89a5-8a3c6c30d752",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Now summarize by number of mutations between wildtype and mutant codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f5c0b88-5ab1-43b9-8bc0-5d9bcf10ac53",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load in wt nucleotide sequence (which is different than the 'wt' sequence from Library as it was codon optimized)\n",
    "niv_m_wt = str(\n",
    "    Bio.SeqIO.read(\n",
    "        \"data/custom_analyses_data/alignments/wild_type_seq.fasta\", \"fasta\"\n",
    "    ).seq\n",
    ")\n",
    "\n",
    "codon_table = {\n",
    "    \"ATA\": \"I\",\n",
    "    \"ATC\": \"I\",\n",
    "    \"ATT\": \"I\",\n",
    "    \"ATG\": \"M\",\n",
    "    \"ACA\": \"T\",\n",
    "    \"ACC\": \"T\",\n",
    "    \"ACG\": \"T\",\n",
    "    \"ACT\": \"T\",\n",
    "    \"AAC\": \"N\",\n",
    "    \"AAT\": \"N\",\n",
    "    \"AAA\": \"K\",\n",
    "    \"AAG\": \"K\",\n",
    "    \"AGC\": \"S\",\n",
    "    \"AGT\": \"S\",\n",
    "    \"AGA\": \"R\",\n",
    "    \"AGG\": \"R\",\n",
    "    \"CTA\": \"L\",\n",
    "    \"CTC\": \"L\",\n",
    "    \"CTG\": \"L\",\n",
    "    \"CTT\": \"L\",\n",
    "    \"CCA\": \"P\",\n",
    "    \"CCC\": \"P\",\n",
    "    \"CCG\": \"P\",\n",
    "    \"CCT\": \"P\",\n",
    "    \"CAC\": \"H\",\n",
    "    \"CAT\": \"H\",\n",
    "    \"CAA\": \"Q\",\n",
    "    \"CAG\": \"Q\",\n",
    "    \"CGA\": \"R\",\n",
    "    \"CGC\": \"R\",\n",
    "    \"CGG\": \"R\",\n",
    "    \"CGT\": \"R\",\n",
    "    \"GTA\": \"V\",\n",
    "    \"GTC\": \"V\",\n",
    "    \"GTG\": \"V\",\n",
    "    \"GTT\": \"V\",\n",
    "    \"GCA\": \"A\",\n",
    "    \"GCC\": \"A\",\n",
    "    \"GCG\": \"A\",\n",
    "    \"GCT\": \"A\",\n",
    "    \"GAC\": \"D\",\n",
    "    \"GAT\": \"D\",\n",
    "    \"GAA\": \"E\",\n",
    "    \"GAG\": \"E\",\n",
    "    \"GGA\": \"G\",\n",
    "    \"GGC\": \"G\",\n",
    "    \"GGG\": \"G\",\n",
    "    \"GGT\": \"G\",\n",
    "    \"TCA\": \"S\",\n",
    "    \"TCC\": \"S\",\n",
    "    \"TCG\": \"S\",\n",
    "    \"TCT\": \"S\",\n",
    "    \"TTC\": \"F\",\n",
    "    \"TTT\": \"F\",\n",
    "    \"TTA\": \"L\",\n",
    "    \"TTG\": \"L\",\n",
    "    \"TAC\": \"Y\",\n",
    "    \"TAT\": \"Y\",\n",
    "    \"TAA\": \"*\",\n",
    "    \"TAG\": \"*\",\n",
    "    \"TGC\": \"C\",\n",
    "    \"TGT\": \"C\",\n",
    "    \"TGA\": \"*\",\n",
    "    \"TGG\": \"W\",\n",
    "}\n",
    "\n",
    "\n",
    "def find_closest_codon(wt_codon, mutant_aa):\n",
    "    mutant_codons = [codon for codon, aa in codon_table.items() if aa == mutant_aa]\n",
    "    min_mutations = 3  # Maximum mutations possible\n",
    "    closest_codon = None\n",
    "    for m_codon in mutant_codons:\n",
    "        mutations = sum([1 for c1, c2 in zip(wt_codon, m_codon) if c1 != c2])\n",
    "        if mutations < min_mutations:\n",
    "            min_mutations = mutations\n",
    "            closest_codon = m_codon\n",
    "    return closest_codon, min_mutations\n",
    "\n",
    "\n",
    "# Function to extract codon for a given site\n",
    "def extract_codon(site):\n",
    "    idx = (site - 1) * 3\n",
    "    return niv_m_wt[idx : idx + 3]\n",
    "\n",
    "\n",
    "def extract_codon_niv_b(site):\n",
    "    idx = (site - 1) * 3\n",
    "    return niv_m_wt[idx : idx + 3]\n",
    "\n",
    "\n",
    "def apply_codon_to_df(df, extract_func):\n",
    "    df[\"wt_codon\"] = df[\"site\"].apply(extract_func)\n",
    "    df[\"closest_mutant_codon\"] = df.apply(\n",
    "        lambda row: find_closest_codon(row[\"wt_codon\"], row[\"mutant\"])[0], axis=1\n",
    "    )\n",
    "    df[\"min_mutations\"] = df.apply(\n",
    "        lambda row: find_closest_codon(row[\"wt_codon\"], row[\"mutant\"])[1], axis=1\n",
    "    )\n",
    "    return df\n",
    "\n",
    "\n",
    "combined_df = apply_codon_to_df(combined_df, extract_codon)\n",
    "filtered_df = apply_codon_to_df(filtered_df, extract_codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ca1f5a8-c826-473c-8dc9-758bb0655d52",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_chart_all(df):\n",
    "    variant_selector = alt.selection_point(\n",
    "        on=\"mouseover\", empty=False, fields=[\"site\"], value=1\n",
    "    )\n",
    "    radio = alt.binding_radio(\n",
    "        options=[1, 2, 3], labels=[\"1\", \"2\", \"3\"], name=\"Min Mutations:\"\n",
    "    )\n",
    "    mutation_selector = alt.param(name=\"MutationSelector\", value=1, bind=radio)\n",
    "\n",
    "    slider = alt.binding_range(min=0.2, max=1.6, step=0.1, name=\"median_escape\")\n",
    "    selector = alt.param(name=\"SelectorName\", value=0.2, bind=slider)\n",
    "\n",
    "    chart = (\n",
    "        alt.Chart(\n",
    "            df,\n",
    "            title=alt.Title(\n",
    "                \"Antibody Escape Mutations\",\n",
    "                subtitle=\"Hover over points to see escape at same site\",\n",
    "            ),\n",
    "        )\n",
    "        .mark_point(filled=True, stroke=\"black\")\n",
    "        .encode(\n",
    "            x=alt.X(\n",
    "                \"ab:O\",\n",
    "                sort=order_ab,\n",
    "                title=\"Antibody\",\n",
    "                axis=alt.Axis(labelAngle=-90, grid=False),\n",
    "            ),\n",
    "            y=alt.Y(\n",
    "                \"effect:Q\",\n",
    "                title=\"Cell Entry of Top Escape\",\n",
    "                axis=alt.Axis(\n",
    "                    grid=True, tickCount=4, values=[0.5, 0, -0.5, -1, -1.5, -2]\n",
    "                ),\n",
    "            ),\n",
    "            size=alt.Size(\n",
    "                \"escape_median\", legend=alt.Legend(title=\"Mean Escape By Mutation\")\n",
    "            ),\n",
    "            xOffset=\"random:Q\",\n",
    "            tooltip=[\n",
    "                \"site\",\n",
    "                \"wildtype\",\n",
    "                \"mutant\",\n",
    "                \"ab\",\n",
    "                \"effect\",\n",
    "                \"escape_median\",\n",
    "                \"escape_std\",\n",
    "            ],\n",
    "            color=alt.Color(\"ab\").legend(None),\n",
    "            opacity=alt.condition(variant_selector, alt.value(1), alt.value(0.4)),\n",
    "            strokeWidth=alt.condition(variant_selector, alt.value(2), alt.value(0)),\n",
    "        )\n",
    "        .transform_calculate(\n",
    "            random=\"sqrt(-1*log(random()))*cos(2*PI*random())\"\n",
    "            # random='random'\n",
    "        )\n",
    "        #.properties(width=config[\"bubble_width\"], height=config[\"bubble_height\"])\n",
    "        .properties(width=200,height=250)\n",
    "        .add_params(variant_selector, mutation_selector, selector)\n",
    "        .transform_filter(\n",
    "            (alt.datum.min_mutations == mutation_selector)\n",
    "            & (alt.datum.escape_median > selector)\n",
    "        )\n",
    "    )\n",
    "\n",
    "    return chart\n",
    "\n",
    "\n",
    "#all_escape = generate_chart_all(combined_df.query(\"escape_median >= 0.2\"))\n",
    "all_escape = generate_chart_all(filtered_df)\n",
    "\n",
    "all_escape.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "962e3c8a-22b5-47ad-b4b2-d918c1aa8b4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make combined figure\n",
    "combined_bubble_plots = (escape_bubble | all_escape)\n",
    "combined_bubble_plots.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e246d7a-a621-46da-8460-c6707606fc56",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def plot_escape_and_mutations_away(df):\n",
    "    variant_selector = alt.selection_point(\n",
    "        on=\"mouseover\", empty=False, fields=[\"site\"], value=1\n",
    "    )\n",
    "    radio = alt.binding_radio(options=[1, 2, 3], name=\"Min Mutations:\")\n",
    "    mutation_selector = alt.param(name=\"MutationSelector\", value=1, bind=radio)\n",
    "\n",
    "    chart = (\n",
    "        alt.Chart(\n",
    "            df,\n",
    "            title=alt.Title(\n",
    "                \"Top Antibody Escape Mutations\",\n",
    "                subtitle=\"By # of nucleotide mutations away\",\n",
    "            ),\n",
    "        )\n",
    "        .mark_point(filled=True, stroke=\"black\")\n",
    "        .encode(\n",
    "            x=alt.X(\n",
    "                \"ab:O\",\n",
    "                sort=order_ab,\n",
    "                title=None,\n",
    "                axis=alt.Axis(labelAngle=-90, grid=False),\n",
    "            ),\n",
    "            y=alt.Y(\n",
    "                \"effect:Q\",\n",
    "                title=\"Cell Entry of Escape Mutants\",\n",
    "                axis=alt.Axis(\n",
    "                    grid=True, tickCount=4, values=[0.5, 0, -0.5, -1, -1.5, -2]\n",
    "                ),\n",
    "            ),  # 'Q' denotes a quantitative variable\n",
    "            size=alt.Size(\"escape_median\", legend=alt.Legend(title=\"Escape of Mutant\")),\n",
    "            xOffset=\"random:Q\",\n",
    "            tooltip=[\"ab\", \"effect\", \"escape_median\", \"site\", \"mutant\"],\n",
    "            opacity=alt.condition(variant_selector, alt.value(1), alt.value(0.4)),\n",
    "            strokeWidth=alt.condition(variant_selector, alt.value(2), alt.value(0)),\n",
    "            color=alt.Color(\"ab:N\").legend(None),\n",
    "        )\n",
    "        .transform_calculate(\n",
    "            # random='random()'\n",
    "            random=\"sqrt(-2*log(random()))*cos(2*PI*random())\"\n",
    "        )\n",
    "        .properties(width=config[\"bubble_width\"], height=config[\"bubble_height\"])\n",
    "        .add_params(variant_selector, mutation_selector)\n",
    "        .transform_filter((alt.datum.min_mutations == mutation_selector))\n",
    "    )\n",
    "    return chart\n",
    "\n",
    "\n",
    "bubble_plot_1_mut_away = plot_escape_and_mutations_away(filtered_df)\n",
    "bubble_plot_1_mut_away.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    bubble_plot_1_mut_away.save(bubble_1_mut_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a07bef14-9b9a-4579-8772-ce6bb0c5759a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_overlapping_escape(df):\n",
    "    slider = alt.binding_range(\n",
    "        min=config[\"min_func_effect_for_ab\"], max=0, step=0.25, name=\"effect\"\n",
    "    )\n",
    "    selector = alt.param(name=\"SelectorName\", value=-4, bind=slider)\n",
    "\n",
    "    radio = alt.binding_radio(options=[1, 2, 3], name=\"Min Mutations:\")\n",
    "    mutation_selector = alt.param(name=\"MutationSelector\", value=1, bind=radio)\n",
    "\n",
    "    df_filtered = df\n",
    "    # Group by 'site' and 'mutant', count the unique 'ab' values for each group\n",
    "    grouped = df_filtered.groupby([\"site\", \"mutant\"])[\"ab\"].nunique().reset_index()\n",
    "\n",
    "    # Filter groups where the count of unique 'ab' values is at least 2\n",
    "    result = grouped[grouped[\"ab\"] >= 2]\n",
    "\n",
    "    # Merge the result with the original dataframe to get the full rows\n",
    "    df_result = pd.merge(df, result[[\"site\", \"mutant\"]], on=[\"site\", \"mutant\"])\n",
    "    df_result[\"mutation_number\"] = (\n",
    "        df_result[\"mutation\"].str.extract(\"(\\d+)\").astype(int)\n",
    "    )\n",
    "    base = (\n",
    "        (\n",
    "            alt.Chart(df_result, title=alt.Title(\"Shared antibody escape mutations\"))\n",
    "            .mark_rect()\n",
    "            .encode(\n",
    "                x=alt.X(\n",
    "                    \"mutation:O\",\n",
    "                    title=\"Site\",\n",
    "                    sort=alt.EncodingSortField(field=\"mutation_number\"),\n",
    "                    axis=alt.Axis(labelAngle=-90, grid=False),\n",
    "                ),\n",
    "                y=alt.Y(\n",
    "                    \"ab:O\", title=\"Mutant\", sort=order_ab, axis=alt.Axis(grid=False)\n",
    "                ),  # Apply custom sort order here\n",
    "                color=\"escape_median\",\n",
    "                tooltip=[\n",
    "                    \"site\",\n",
    "                    \"wildtype\",\n",
    "                    \"mutant\",\n",
    "                    \"escape_median\",\n",
    "                    \"min_mutations\",\n",
    "                ],\n",
    "            )\n",
    "        )\n",
    "        .properties(height=200, width=400)\n",
    "        .add_params(selector, mutation_selector)\n",
    "        .transform_filter(\n",
    "            (alt.datum.effect >= selector)\n",
    "            & (alt.datum.min_mutations == mutation_selector)\n",
    "        )\n",
    "    )\n",
    "    return base\n",
    "\n",
    "\n",
    "overlap_escape = find_overlapping_escape(filtered_df)\n",
    "overlap_escape.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    overlap_escape.save(overlap_escape_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55a08d61-96a5-423c-a0f6-edd7aed61cff",
   "metadata": {},
   "source": [
    "### Line plots of escape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50556cec-9298-441b-b9ca-621777ac2427",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_line_escape(df):\n",
    "    variant_selector = alt.selection_point(\n",
    "        on=\"mouseover\", empty=False, fields=[\"site\"], value=0\n",
    "    )\n",
    "    # Group by 'site' and 'mutant', count the unique 'ab' values for each group\n",
    "    summed = df.groupby([\"site\", \"ab\"])[\"escape_median\"].sum().reset_index()\n",
    "    empty_chart = []\n",
    "    ab_list = [\"m102.4\", \"HENV-26\", \"HENV-117\", \"HENV-103\", \"HENV-32\", \"nAH1.3\"]\n",
    "    for idx, ab in enumerate(ab_list):\n",
    "        tmp_df = summed[summed[\"ab\"] == ab]\n",
    "        # color = '#1f4e79'\n",
    "        if ab in [\"m102.4\", \"HENV-26\", \"HENV-117\"]:\n",
    "            color = \"#1f4e79\"\n",
    "        if ab in [\"HENV-103\", \"HENV-32\"]:\n",
    "            color = \"#ff7f0e\"\n",
    "        if ab in [\"nAH1.3\"]:\n",
    "            color = \"#2ca02c\"\n",
    "\n",
    "        # Conditionally set the x-axis labels and title for the last plot\n",
    "        is_last_plot = idx == len(ab_list) - 1\n",
    "        x_axis = alt.Axis(\n",
    "            values=[100, 200, 300, 400, 500, 600],\n",
    "            tickCount=6,\n",
    "            labelAngle=-90,\n",
    "            grid=True,\n",
    "            labelExpr=\"datum.value % 100 === 0 ? datum.value : ''\",\n",
    "            title=\"Site\" if is_last_plot else None,\n",
    "            labels=is_last_plot,\n",
    "        )  # Only show labels for the last plot\n",
    "        base = (\n",
    "            alt.Chart(tmp_df)\n",
    "            .mark_line(size=1, color=color)\n",
    "            .encode(\n",
    "                x=alt.X(\"site:O\", axis=x_axis),\n",
    "                y=alt.Y(\"escape_median\", title=f\"{ab}\", axis=alt.Axis(grid=True)),\n",
    "            )\n",
    "            .properties(\n",
    "                width=config[\"large_line_width\"], height=config[\"large_line_height\"]\n",
    "            )\n",
    "        )\n",
    "        point = (\n",
    "            base.mark_point(color=\"black\", size=10, filled=True)\n",
    "            .encode(\n",
    "                x=alt.X(\"site:O\", axis=x_axis),\n",
    "                y=alt.Y(\n",
    "                    \"escape_median\",\n",
    "                    title=f\"{ab}\",\n",
    "                    axis=alt.Axis(\n",
    "                        grid=True,\n",
    "                    ),\n",
    "                ),\n",
    "                size=alt.condition(variant_selector, alt.value(100), alt.value(15)),\n",
    "                color=alt.condition(\n",
    "                    variant_selector, alt.value(\"black\"), alt.value(color)\n",
    "                ),\n",
    "                tooltip=[\"site\", \"escape_median\"],\n",
    "            )\n",
    "            .add_params(variant_selector)\n",
    "        )\n",
    "        chart = base + point\n",
    "        empty_chart.append(chart)\n",
    "\n",
    "    # Use configure_concat to adjust spacing between vertically concatenated plots\n",
    "    combined_chart = (\n",
    "        alt.vconcat(*empty_chart, spacing=1)\n",
    "        .resolve_scale(y=\"independent\", x=\"shared\", color=\"independent\")\n",
    "        .properties(\n",
    "            title=alt.Title(\n",
    "                \"Summed Antibody Escape by Site\", subtitle=\"Colored by epitope\"\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "\n",
    "    return combined_chart\n",
    "\n",
    "\n",
    "tmp_line = plot_line_escape(combined_df)\n",
    "tmp_line.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    tmp_line.save(mab_line_escape_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df646b0d-1e3b-494f-924b-83f95af8ddec",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Now calculate atomic distances between escape sites and closest amino acid in heavy and light chains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7690121-ac4d-4ac7-b66d-5554c38773b8",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def calculate_min_distances(pdb_path, source_chain_id, target_chain_ids, name):\n",
    "    # Initialize the PDB parser and load the structure\n",
    "    parser = PDB.PDBParser(QUIET=True)\n",
    "    structure = parser.get_structure(\"structure_id\", pdb_path)\n",
    "\n",
    "    source_chain = structure[0][source_chain_id]\n",
    "    target_chains = [structure[0][chain_id] for chain_id in target_chain_ids]\n",
    "\n",
    "    data = []\n",
    "\n",
    "    for residueA in source_chain:\n",
    "        if residueA.resname in [\"HOH\", \"WAT\", \"IPA\", \"NAG\"]:\n",
    "            continue\n",
    "\n",
    "        min_distance = float(\"inf\")\n",
    "        closest_residueB = None\n",
    "        closest_chain_id = None\n",
    "        residues_within_4 = 0\n",
    "\n",
    "        for target_chain in target_chains:\n",
    "            for residueB in target_chain:\n",
    "                if residueB.resname in [\"HOH\", \"WAT\", \"IPA\"]:\n",
    "                    continue\n",
    "\n",
    "                # Check for residues within 4 angstroms\n",
    "                is_within_4 = False\n",
    "                for atomA in residueA:\n",
    "                    for atomB in residueB:\n",
    "                        distance = atomA - atomB\n",
    "                        if distance < min_distance:\n",
    "                            min_distance = distance\n",
    "                            closest_residueB = residueB\n",
    "                            closest_chain_id = target_chain.get_id()\n",
    "                        if distance < 4:\n",
    "                            is_within_4 = True\n",
    "                if is_within_4:\n",
    "                    residues_within_4 += 1\n",
    "\n",
    "        data.append(\n",
    "            {\n",
    "                \"wildtype\": residueA.resname,\n",
    "                \"site\": residueA.id[1],\n",
    "                \"chain\": closest_chain_id,\n",
    "                \"residue\": closest_residueB.id[1],\n",
    "                \"residue_name\": closest_residueB.resname,\n",
    "                \"distance\": min_distance,\n",
    "                \"residues_within_4\": residues_within_4,\n",
    "                \"ab\": name,\n",
    "            }\n",
    "        )\n",
    "\n",
    "    # Convert data to pandas DataFrame\n",
    "    df = pd.DataFrame(data)\n",
    "    return df\n",
    "\n",
    "\n",
    "def check_file(input_path, source_chain, target_chain, name, output_path):\n",
    "\n",
    "    file_path = output_path\n",
    "\n",
    "    if not os.path.exists(file_path):\n",
    "        print(f\"File {name} does not exist, running calculation\")\n",
    "        output_df = calculate_min_distances(\n",
    "            input_path, source_chain, target_chain, name\n",
    "        )\n",
    "        print(f\"done calculating for {file_path}\")\n",
    "        output_df.to_csv(output_path, index=False)\n",
    "        return output_df\n",
    "    else:\n",
    "        print(\"File already exists,loading from disk\")\n",
    "        output_df = pd.read_csv(output_path)\n",
    "        return output_df\n",
    "\n",
    "\n",
    "pdb_path_26 = \"data/custom_analyses_data/crystal_structures/6vy5.pdb\"\n",
    "source_chain_26 = \"A\"\n",
    "target_chains_26 = [\"H\", \"L\"]\n",
    "output_path_26 = \"results/distances/df_HENV26_atomic_distances.csv\"\n",
    "\n",
    "pdb_path_32 = \"data/custom_analyses_data/crystal_structures/6vy4.pdb\"\n",
    "source_chain_32 = \"A\"\n",
    "target_chains_32 = [\"H\", \"L\"]\n",
    "output_path_32 = \"results/distances/df_HENV32_atomic_distances.csv\"\n",
    "\n",
    "pdb_path_nah = \"data/custom_analyses_data/crystal_structures/7txz.pdb\"\n",
    "source_chain_nah = \"A\"\n",
    "target_chains_nah = [\"F\", \"E\"]\n",
    "output_path_nah = \"results/distances/df_nAH_atomic_distances.csv\"\n",
    "\n",
    "pdb_path_m102 = \"data/custom_analyses_data/crystal_structures/6cmg.pdb\"\n",
    "source_chain_m102 = \"A\"\n",
    "target_chains_m102 = [\"B\", \"C\"]\n",
    "output_path_m102 = \"results/distances/df_m102_atomic_distances.csv\"\n",
    "\n",
    "\n",
    "df_HENV26 = check_file(\n",
    "    pdb_path_26, source_chain_26, target_chains_26, \"HENV-26\", output_path_26\n",
    ")\n",
    "df_HENV32 = check_file(\n",
    "    pdb_path_32, source_chain_32, target_chains_32, \"HENV-32\", output_path_32\n",
    ")\n",
    "df_nah = check_file(\n",
    "    pdb_path_nah, source_chain_nah, target_chains_nah, \"nAH1.3\", output_path_nah\n",
    ")\n",
    "df_nah[\"chain\"].replace(\n",
    "    {\"E\": \"H\", \"F\": \"L\"}, inplace=True\n",
    ")  # Fix naming so consistent heavy and light chain naming\n",
    "df_m102 = check_file(\n",
    "    pdb_path_m102, source_chain_m102, target_chains_m102, \"m102.4\", output_path_m102\n",
    ")\n",
    "df_m102[\"chain\"].replace(\n",
    "    {\"C\": \"H\", \"B\": \"L\"}, inplace=True\n",
    ")  # Fix naming so consistent heavy and light chain naming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6251849-9b43-445d-ac6e-b1c9c6f3ae2b",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_close_mab_sites(df, name):\n",
    "    unique_sites = df.query(\"distance <= 4\")[\"site\"].unique()\n",
    "    mab_site_list = list(unique_sites)\n",
    "    print(f\"Close sites for mAb {name} are: {mab_site_list}\")\n",
    "    return mab_site_list\n",
    "\n",
    "\n",
    "### First find RBP sites that are close to mAb residues\n",
    "nah_close = find_close_mab_sites(df_nah, \"nAH1.3\")\n",
    "HENV26_close = find_close_mab_sites(df_HENV26, \"HENV-26\")\n",
    "HENV32_close = find_close_mab_sites(df_HENV32, \"HENV-32\")\n",
    "m102_close = find_close_mab_sites(df_m102, \"m102.4\")\n",
    "\n",
    "### Now combined the close residues AND the top escape sites identified previously\n",
    "nah_combined_sites = sites_dict[\"nAH1.3\"] + nah_close\n",
    "HENV26_combined_sites = sites_dict[\"HENV-26\"] + HENV26_close\n",
    "HENV32_combined_sites = sites_dict[\"HENV-32\"] + HENV32_close\n",
    "m102_combined_sites = sites_dict[\"m102.4\"] + m102_close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4daeeaf-7cc0-42fe-8e36-afb2ecf3a68c",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def make_distance(df):\n",
    "    subset_df = df[[\"site\", \"distance\"]].copy()\n",
    "    subset_df[\"mutant\"] = \"distance\"\n",
    "    subset_df[\"wildtype\"] = \"\"\n",
    "    subset_df[\"effect\"] = \"escape_median\"\n",
    "    subset_df.rename(columns={\"distance\": \"value\"}, inplace=True)\n",
    "    return subset_df\n",
    "\n",
    "\n",
    "distance_nah_df = make_distance(df_nah)\n",
    "distance_26_df = make_distance(df_HENV26)\n",
    "distance_32_df = make_distance(df_HENV32)\n",
    "distance_m102_df = make_distance(df_m102)\n",
    "\n",
    "display(distance_m102_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31e90b09-71d1-4c58-a39d-e525e7b8bb40",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Prepare dataframes for heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a64a6be3-2886-471f-9cfb-d0872ccbe1b6",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def make_empty_df_with_distance(ab, distance_file):\n",
    "    # print(ab)\n",
    "    sites = range(71, 603)\n",
    "    amino_acids = [\n",
    "        \"A\",\n",
    "        \"C\",\n",
    "        \"D\",\n",
    "        \"E\",\n",
    "        \"F\",\n",
    "        \"G\",\n",
    "        \"H\",\n",
    "        \"I\",\n",
    "        \"K\",\n",
    "        \"L\",\n",
    "        \"M\",\n",
    "        \"N\",\n",
    "        \"P\",\n",
    "        \"Q\",\n",
    "        \"R\",\n",
    "        \"S\",\n",
    "        \"T\",\n",
    "        \"V\",\n",
    "        \"W\",\n",
    "        \"Y\",\n",
    "    ]\n",
    "    # Create the combination of each site with each amino acid\n",
    "    data = [{\"site\": site, \"mutant\": aa} for site in sites for aa in amino_acids]\n",
    "    # Create the DataFrame\n",
    "    empty_df = pd.DataFrame(data)\n",
    "    all_sites_df = pd.merge(\n",
    "        empty_df, combined_df.query(f'ab == \"{ab}\"'), on=[\"site\", \"mutant\"], how=\"left\"\n",
    "    )\n",
    "    df_melted = all_sites_df.melt(\n",
    "        id_vars=[\"site\", \"mutant\", \"wildtype\"],\n",
    "        value_vars=[\"escape_median\"],\n",
    "        var_name=\"effect\",\n",
    "        value_name=\"value\",\n",
    "    )\n",
    "\n",
    "    df_filtered = func_scores_E3_low_effect.melt(\n",
    "        id_vars=[\"site\", \"mutant\", \"wildtype\"],\n",
    "        value_vars=[\"effect\"],\n",
    "        var_name=\"effect\",\n",
    "        value_name=\"value\",\n",
    "    )\n",
    "\n",
    "    df_test = pd.concat([df_melted, df_filtered, distance_file], ignore_index=True)\n",
    "    df_test[\"ab\"] = ab\n",
    "    return df_test\n",
    "\n",
    "\n",
    "empty_df_m102 = make_empty_df_with_distance(\"m102.4\", distance_m102_df)\n",
    "empty_df_HENV26 = make_empty_df_with_distance(\"HENV-26\", distance_26_df)\n",
    "empty_df_HENV32 = make_empty_df_with_distance(\"HENV-32\", distance_32_df)\n",
    "empty_df_nah = make_empty_df_with_distance(\"nAH1.3\", distance_nah_df)\n",
    "\n",
    "\n",
    "def make_empty_df(ab):\n",
    "    sites = range(71, 603)\n",
    "    amino_acids = [\n",
    "        \"A\",\n",
    "        \"C\",\n",
    "        \"D\",\n",
    "        \"E\",\n",
    "        \"F\",\n",
    "        \"G\",\n",
    "        \"H\",\n",
    "        \"I\",\n",
    "        \"K\",\n",
    "        \"L\",\n",
    "        \"M\",\n",
    "        \"N\",\n",
    "        \"P\",\n",
    "        \"Q\",\n",
    "        \"R\",\n",
    "        \"S\",\n",
    "        \"T\",\n",
    "        \"V\",\n",
    "        \"W\",\n",
    "        \"Y\",\n",
    "    ]\n",
    "    # Create the combination of each site with each amino acid\n",
    "    data = [{\"site\": site, \"mutant\": aa} for site in sites for aa in amino_acids]\n",
    "    # Create the DataFrame\n",
    "    empty_df = pd.DataFrame(data)\n",
    "    all_sites_df = pd.merge(\n",
    "        empty_df, combined_df.query(f'ab == \"{ab}\"'), on=[\"site\", \"mutant\"], how=\"left\"\n",
    "    )\n",
    "    df_melted = all_sites_df.melt(\n",
    "        id_vars=[\"site\", \"mutant\", \"wildtype\"],\n",
    "        value_vars=[\"escape_median\"],\n",
    "        var_name=\"effect\",\n",
    "        value_name=\"value\",\n",
    "    )\n",
    "\n",
    "    df_filtered = func_scores_E3_low_effect.melt(\n",
    "        id_vars=[\"site\", \"mutant\", \"wildtype\"],\n",
    "        value_vars=[\"effect\"],\n",
    "        var_name=\"effect\",\n",
    "        value_name=\"value\",\n",
    "    )\n",
    "\n",
    "    df_test = pd.concat([df_melted, df_filtered], ignore_index=True)\n",
    "    df_test[\"ab\"] = ab\n",
    "    return df_test\n",
    "\n",
    "\n",
    "empty_df_HENV117 = make_empty_df(\"HENV-117\")\n",
    "empty_df_HENV103 = make_empty_df(\"HENV-103\")\n",
    "\n",
    "combined_ab = pd.concat(\n",
    "    [\n",
    "        empty_df_m102,\n",
    "        empty_df_HENV26,\n",
    "        empty_df_HENV32,\n",
    "        empty_df_nah,\n",
    "        empty_df_HENV117,\n",
    "        empty_df_HENV103,\n",
    "    ]\n",
    ")\n",
    "display(combined_ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9016f96-d629-4102-971a-ddefdb7c9e48",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def plot_distance_only(df, trigger):\n",
    "    custom_order = [\n",
    "        \"distance\",\n",
    "        \"R\",\n",
    "        \"K\",\n",
    "        \"H\",\n",
    "        \"D\",\n",
    "        \"E\",\n",
    "        \"Q\",\n",
    "        \"N\",\n",
    "        \"S\",\n",
    "        \"T\",\n",
    "        \"Y\",\n",
    "        \"W\",\n",
    "        \"F\",\n",
    "        \"A\",\n",
    "        \"I\",\n",
    "        \"L\",\n",
    "        \"M\",\n",
    "        \"V\",\n",
    "        \"G\",\n",
    "        \"P\",\n",
    "        \"C\",\n",
    "    ]\n",
    "    all_residues = range(71, 603)\n",
    "    final_df = df\n",
    "    final_df = final_df.sort_values(\n",
    "        \"site\"\n",
    "    )  # Sort the dataframe by 'site' to ensure that duplicates are detected correctly.\n",
    "    sort_order = {\n",
    "        mutant: i for i, mutant in enumerate(custom_order)\n",
    "    }  # Create a dictionary that maps each mutant to its sort rank based on the custom order\n",
    "    final_df[\"mutant_rank\"] = final_df[\"mutant\"].map(\n",
    "        sort_order\n",
    "    )  # Map the 'mutant' column to these ranks\n",
    "\n",
    "    final_df = final_df.sort_values(\n",
    "        \"mutant_rank\"\n",
    "    )  # Now sort the dataframe by this rank\n",
    "    final_df = final_df.drop(\n",
    "        columns=[\"mutant_rank\"]\n",
    "    )  # Drop the 'mutant_rank' column as it is no longer needed after sorting\n",
    "    sites = sorted(final_df[\"site\"].unique(), key=lambda x: float(x))\n",
    "    ab_list = [\"m102.4\", \"HENV-26\", \"HENV-117\", \"HENV-103\", \"HENV-32\", \"nAH1.3\"]\n",
    "    empty_chart = []  # setup collection for charts\n",
    "    for idx, ab in enumerate(ab_list):\n",
    "        tmp_df = final_df[final_df[\"ab\"] == ab]\n",
    "        if ab == \"m102.4\":\n",
    "            site_subset = m102_combined_sites\n",
    "            # legend_conditional = alt.Legend(title='Distance to mAb')\n",
    "        if ab == \"HENV-26\":\n",
    "            site_subset = HENV26_combined_sites\n",
    "            # legend_conditional = alt.Legend(title='Distance to mAb')\n",
    "        if ab == \"HENV-32\":\n",
    "            site_subset = HENV32_combined_sites\n",
    "            # legend_conditional = alt.Legend(title='Distance to mAb')\n",
    "        if ab == \"HENV-103\":\n",
    "            site_subset = sites_dict[\"HENV-103\"]\n",
    "            # legend_conditional = alt.Legend(title=None)\n",
    "        if ab == \"HENV-117\":\n",
    "            site_subset = sites_dict[\"HENV-117\"]\n",
    "            # legend_conditional = alt.Legend(title=None)\n",
    "        if ab == \"nAH1.3\":\n",
    "            site_subset = nah_combined_sites\n",
    "            # legend_conditional = alt.Legend(title='Distance to mAb')\n",
    "\n",
    "        # select which sites you will show\n",
    "        if trigger == True:\n",
    "            tmp_df = tmp_df[tmp_df[\"site\"].isin(site_subset)]\n",
    "            x_axis = alt.Axis(\n",
    "                labelAngle=-90,\n",
    "                # labelExpr=\"datum.value % 10 === 0 ? datum.value : ''\",\n",
    "                title=\"Site\",\n",
    "            )\n",
    "        else:\n",
    "            tmp_df = tmp_df[tmp_df[\"site\"].isin(all_residues)]\n",
    "\n",
    "            # Conditionally set the x-axis labels and title for the last plot\n",
    "            is_last_plot = idx == len(ab_list) - 1\n",
    "            x_axis = alt.Axis(\n",
    "                labelAngle=-90,\n",
    "                labelExpr=\"datum.value % 10 === 0 ? datum.value : ''\",\n",
    "                title=\"Site\" if is_last_plot else None,\n",
    "                labels=True,\n",
    "            )  # Only show labels for the last plot\n",
    "\n",
    "        # Prepare the color scales separately for distance and effects\n",
    "        # Filter out 'distance' values before creating the effect heatmap\n",
    "        effect_df = tmp_df[\n",
    "            (tmp_df[\"mutant\"] != \"distance\") & (tmp_df[\"effect\"] != \"effect\")\n",
    "        ]\n",
    "        max_color = effect_df[\"value\"].max()\n",
    "        min_color = effect_df[\"value\"].min()\n",
    "\n",
    "        # Adjust color scheme for abs with little sensitizing mutations\n",
    "        if min_color > -1:\n",
    "            min_color = min_color - 1\n",
    "\n",
    "        # Prepare the color scale for effects, Altair will automatically determine the domain\n",
    "        color_scale_escape = alt.Scale(\n",
    "            scheme=\"redblue\", domainMid=0, domain=[min_color, max_color]\n",
    "        )\n",
    "        color_scale_entropy = alt.Scale(scheme=\"greens\", domain=[0, 15], reverse=True)\n",
    "\n",
    "        strokewidth_size = 0.25\n",
    "\n",
    "        unique_wildtypes_df = tmp_df.drop_duplicates(subset=[\"site\", \"wildtype\"])\n",
    "\n",
    "        # The chart for the heatmap\n",
    "        base = (\n",
    "            alt.Chart(tmp_df, title=f\"{ab}\")\n",
    "            .encode(\n",
    "                x=alt.X(\"site:O\", title=\"Site\", sort=sites, axis=x_axis),\n",
    "                y=alt.Y(\n",
    "                    \"mutant\",\n",
    "                    title=\"Amino Acid\",\n",
    "                    sort=alt.EncodingSortField(field=\"sort_order\", order=\"ascending\"),\n",
    "                    axis=alt.Axis(grid=False),\n",
    "                ),  # Apply custom sort order here\n",
    "                tooltip=[\"site\", \"wildtype\", \"mutant\", \"value\"],\n",
    "            )\n",
    "            .properties(width=alt.Step(10), height=alt.Step(11))\n",
    "        )\n",
    "        # Heatmap for distance\n",
    "        chart_empty = (\n",
    "            base.mark_rect(color=\"#e6e7e8\")\n",
    "            .encode()\n",
    "            .transform_filter(alt.datum.effect == \"escape_median\")\n",
    "        )\n",
    "        # Heatmap for effect\n",
    "        chart_effect = (\n",
    "            base.mark_rect(stroke=\"black\", strokeWidth=strokewidth_size)\n",
    "            .encode(\n",
    "                color=alt.condition(\n",
    "                    'datum.mutant != \"distance\"',\n",
    "                    alt.Color(\n",
    "                        \"value:Q\",\n",
    "                        scale=color_scale_escape,\n",
    "                        legend=alt.Legend(title=f\"{ab} Escape\"),\n",
    "                    ),\n",
    "                    alt.value(\"transparent\"),\n",
    "                ),\n",
    "            )\n",
    "            .transform_filter(alt.datum.effect == \"escape_median\")\n",
    "        )\n",
    "\n",
    "        # Heatmap for distance\n",
    "        if ab in [\"m102.4\", \"HENV-26\", \"HENV-32\", \"nAH1.3\"]:\n",
    "            chart_distance = (\n",
    "                base.mark_rect()\n",
    "                .encode(\n",
    "                    color=alt.condition(\n",
    "                        'datum.mutant == \"distance\"',\n",
    "                        alt.Color(\n",
    "                            \"value:Q\",\n",
    "                            scale=color_scale_entropy,\n",
    "                            legend=alt.Legend(title=\"Distance to mAb\"),\n",
    "                        ),\n",
    "                        alt.value(\"transparent\"),\n",
    "                    )\n",
    "                )\n",
    "                .transform_filter(alt.datum.effect == \"escape_median\")\n",
    "            )\n",
    "        else:\n",
    "            chart_distance = (\n",
    "                base.mark_rect(color=\"transparent\")\n",
    "                .encode(\n",
    "                    # color=alt.Color('white'),\n",
    "                    # alt.Color('value:Q', scale=color_scale_entropy,legend=alt.Legend(title='Distance to mAb')),\n",
    "                    # alt.value('transparent'))\n",
    "                )\n",
    "                .transform_filter(alt.datum.effect == \"escape_median\")\n",
    "            )\n",
    "        # Heatmap for distance\n",
    "        chart_filtered = (\n",
    "            base.mark_rect(\n",
    "                color=\"#939598\", stroke=\"black\", strokeWidth=strokewidth_size\n",
    "            )\n",
    "            .encode()\n",
    "            .transform_filter(alt.datum.effect == \"effect\")\n",
    "        )\n",
    "\n",
    "        # The layer for the wildtype boxes\n",
    "        wildtype_layer_box = (\n",
    "            alt.Chart(unique_wildtypes_df)\n",
    "            .mark_rect(color=\"white\", stroke=\"black\", strokeWidth=strokewidth_size)\n",
    "            .encode(\n",
    "                x=alt.X(\"site:O\", sort=sites),\n",
    "                y=alt.Y(\n",
    "                    \"wildtype\",\n",
    "                    sort=alt.EncodingSortField(field=\"sort_order\", order=\"ascending\"),\n",
    "                ),\n",
    "                opacity=alt.value(1),\n",
    "            )\n",
    "            .transform_filter(\n",
    "                (alt.datum.wildtype != \"\")\n",
    "                & (alt.datum.wildtype != None)\n",
    "                & (alt.datum.value != None)\n",
    "            )\n",
    "        )\n",
    "        # The layer for the wildtype amino acids\n",
    "        wildtype_layer = (\n",
    "            alt.Chart(unique_wildtypes_df)\n",
    "            .mark_text(color=\"black\", text=\"X\", size=8)\n",
    "            .encode(\n",
    "                x=alt.X(\"site:O\", sort=sites),\n",
    "                y=alt.Y(\n",
    "                    \"wildtype\",\n",
    "                    sort=alt.EncodingSortField(field=\"sort_order\", order=\"ascending\"),\n",
    "                ),\n",
    "                opacity=alt.value(1),\n",
    "            )\n",
    "            .transform_filter(\n",
    "                (alt.datum.wildtype != \"\")\n",
    "                & (alt.datum.wildtype != None)\n",
    "                & (alt.datum.value != None)\n",
    "            )\n",
    "        )\n",
    "\n",
    "        # Combine the heatmap layer with the wildtype layer\n",
    "        chart = alt.layer(\n",
    "            chart_empty,\n",
    "            chart_effect,\n",
    "            chart_distance,\n",
    "            chart_filtered,\n",
    "            wildtype_layer_box,\n",
    "            wildtype_layer,\n",
    "        ).resolve_scale(color=\"independent\")\n",
    "        empty_chart.append(chart)\n",
    "    combined_chart = (\n",
    "        alt.vconcat(*empty_chart, spacing=1)\n",
    "        .resolve_scale(y=\"shared\", x=\"independent\", color=\"independent\")\n",
    "        .configure_title(\n",
    "            anchor=\"start\",  # Aligns the title to the left ('middle' for center, 'end' for right)\n",
    "            offset=10,  # Adjusts the distance of the title from the chart\n",
    "            orient=\"top\",  # Positions the title at the top; use 'bottom' to position at the bottom\n",
    "        )\n",
    "    )\n",
    "    return combined_chart\n",
    "\n",
    "\n",
    "mab_plot = plot_distance_only(combined_ab, True)\n",
    "mab_plot.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    mab_plot.save(mab_plot_top)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c5204ed-a60f-471f-b8e1-471d9bd1b6cb",
   "metadata": {},
   "source": [
    "### Make full antibody escape heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8039f1f7-fede-43ff-bf0b-337e18a8d2eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mab_all = plot_distance_only(combined_ab, False)\n",
    "mab_all.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    mab_all.save(mab_plot_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d251055-d7c2-4e0c-adce-0ded6829d161",
   "metadata": {},
   "source": [
    "### Now make heatmaps of antibody escape versus Ephrin Binding"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d3f4560-3377-425b-a171-abe21d5626bb",
   "metadata": {},
   "source": [
    "First prepare data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dee789e-c24e-4ec1-9442-c7f83db211d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "bind_df = pd.read_csv(binding_data)\n",
    "binding_df = bind_df.groupby(\"site\")[\"binding_median\"].median().reset_index()\n",
    "\n",
    "\n",
    "def make_empty_binding():\n",
    "    sites = range(71, 603)\n",
    "    data = [{\"site\": site} for site in sites]\n",
    "    empty_df = pd.DataFrame(data)\n",
    "    empty_df = pd.merge(empty_df, binding_df, on=\"site\", how=\"left\")\n",
    "    empty_df = empty_df.rename(columns={\"binding_median\": \"value\"})\n",
    "    empty_df[\"effect\"] = \"escape_median\"\n",
    "    empty_df[\"ab\"] = \"Ephrin-B2 binding\"\n",
    "    return empty_df\n",
    "\n",
    "\n",
    "binding_empty = make_empty_binding()\n",
    "\n",
    "escape_df = combined_df.groupby([\"ab\", \"site\"])[\"escape_median\"].median().reset_index()\n",
    "\n",
    "\n",
    "def make_empty_df(ab):\n",
    "    sites = range(71, 603)\n",
    "    data = [{\"site\": site} for site in sites]\n",
    "\n",
    "    # Create the DataFrame\n",
    "    empty_df = pd.DataFrame(data)\n",
    "\n",
    "    all_sites_df = pd.merge(\n",
    "        empty_df, escape_df.query(f'ab == \"{ab}\"'), on=[\"site\"], how=\"left\"\n",
    "    )\n",
    "\n",
    "    df_melted = all_sites_df.melt(\n",
    "        id_vars=[\"site\"],\n",
    "        value_vars=[\"escape_median\"],\n",
    "        var_name=\"effect\",\n",
    "        value_name=\"value\",\n",
    "    )\n",
    "\n",
    "    df_test = pd.concat([df_melted], ignore_index=True)\n",
    "    df_test[\"ab\"] = ab\n",
    "    return df_test\n",
    "\n",
    "\n",
    "ab_list = [\"m102.4\", \"HENV-26\", \"HENV-117\", \"HENV-103\", \"HENV-32\", \"nAH1.3\"]\n",
    "# ab_list = ['HENV-32']\n",
    "\n",
    "empty = []\n",
    "for ab in ab_list:\n",
    "    tmp_df = make_empty_df(ab)\n",
    "    empty.append(tmp_df)\n",
    "all_empties_df = pd.concat(empty, ignore_index=True)\n",
    "all_empties_df = pd.concat([all_empties_df, binding_empty])\n",
    "display(all_empties_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1f17cb9-79a7-4163-a1e6-ddafaa9bda11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_heatmap_with_binding(df):\n",
    "    # Define the custom sort order directly in the encoding\n",
    "    sort_order = [\n",
    "        \"NiV Polymorphism\",\n",
    "        \"Ephrin-B2 binding\",\n",
    "        \"m102.4\",\n",
    "        \"HENV-26\",\n",
    "        \"HENV-117\",\n",
    "        \"HENV-103\",\n",
    "        \"HENV-32\",\n",
    "        \"nAH1.3\",\n",
    "    ]\n",
    "    full_ranges = [\n",
    "        list(range(start, end))\n",
    "        for start, end in [(71, 181), (181, 291), (291, 401), (401, 511), (511, 603)]\n",
    "    ]\n",
    "\n",
    "    # container to hold the charts\n",
    "    charts = []\n",
    "    color_scale_effect = alt.Scale(scheme=\"redblue\", domainMid=0)\n",
    "    color_scale_binding = alt.Scale(scheme=\"redblue\", domainMid=0)\n",
    "\n",
    "    for idx, subset in enumerate(full_ranges):\n",
    "        subset_df = df[df[\"site\"].isin(subset)]  # for the wrapping of sites\n",
    "        is_last_plot = idx == len(full_ranges) - 1\n",
    "        x_axis = alt.Axis(\n",
    "            labelAngle=-90,\n",
    "            labelExpr=\"datum.value % 10 === 0 ? datum.value : ''\",\n",
    "            title=\"Site\" if is_last_plot else None,\n",
    "            labels=True,\n",
    "        )  # Only show labels for the last plot\n",
    "\n",
    "        effect_legend = (\n",
    "            alt.Legend(title=\"Antibody Escape\") if is_last_plot else None\n",
    "        )  # ,direction='horizontal',gradientLength=50,titleAnchor='middle',tickCount=3,labelAlign='center')\n",
    "        binding_legend = (\n",
    "            alt.Legend(title=\"Henipavirus Entropy\") if is_last_plot else None\n",
    "        )  # ,direction='horizontal',gradientLength=50,titleAnchor='middle',labelAlign='center')\n",
    "        print(is_last_plot)\n",
    "        print(effect_legend)\n",
    "        base = (\n",
    "            alt.Chart(subset_df)\n",
    "            .encode(\n",
    "                x=alt.X(\"site:O\", title=\"Site\", axis=x_axis),\n",
    "                y=alt.Y(\n",
    "                    \"ab\", title=None, sort=sort_order, axis=alt.Axis(grid=False)\n",
    "                ),  # Correctly apply custom sort order\n",
    "                tooltip=[\"site\", \"value\"],\n",
    "            )\n",
    "            .properties(width=alt.Step(10), height=alt.Step(11))\n",
    "        )\n",
    "\n",
    "        # Define the chart for empty cells\n",
    "        chart_empty = base.mark_rect(color=\"#e6e7e8\").transform_filter(\n",
    "            alt.datum.effect == \"escape_median\"\n",
    "        )\n",
    "\n",
    "        # Define the chart for cells with effect\n",
    "        chart_effect = (\n",
    "            base.mark_rect(stroke=\"black\", strokeWidth=0.25)\n",
    "            .encode(\n",
    "                color=alt.condition(\n",
    "                    'datum.effect == \"escape_median\"',\n",
    "                    alt.Color(\n",
    "                        \"value:Q\", scale=color_scale_effect, legend=effect_legend\n",
    "                    ),  # Define a color scale\n",
    "                    alt.value(\"transparent\"),\n",
    "                )\n",
    "            )\n",
    "            .transform_filter(alt.datum.effect == \"escape_median\")\n",
    "        )\n",
    "\n",
    "        chart_binding = (\n",
    "            base.mark_rect(strokeWidth=1.1)\n",
    "            .encode(\n",
    "                stroke=alt.value(\"value\"),\n",
    "                color=alt.condition(\n",
    "                    'datum.effect == \"escape_median\"',\n",
    "                    alt.Color(\n",
    "                        \"value:Q\", scale=color_scale_binding, legend=binding_legend\n",
    "                    ),\n",
    "                    alt.value(\"transparent\"),\n",
    "                ),\n",
    "            )\n",
    "            .transform_filter(alt.datum.ab == \"Ephrin-B2 binding\")\n",
    "        )\n",
    "\n",
    "        chart_poly = (\n",
    "            base.mark_rect(color=\"black\")\n",
    "            .encode()\n",
    "            .transform_filter(alt.datum.ab == \"NiV Polymorphism\")\n",
    "        )\n",
    "        # Layer the charts using `layer` instead of `+`\n",
    "        chart = alt.layer(\n",
    "            chart_empty, chart_effect, chart_binding, chart_poly\n",
    "        )  # .resolve_scale(color='shared')\n",
    "        charts.append(chart)\n",
    "    combined_chart = alt.vconcat(\n",
    "        *charts, spacing=5, title=\"Heatmap of median mAb escape and Ephrin-B2 binding\"\n",
    "    )  \n",
    "\n",
    "    return combined_chart\n",
    "\n",
    "\n",
    "# Assuming `all_empties_df` is your DataFrame and already defined\n",
    "chart = make_heatmap_with_binding(all_empties_df)\n",
    "chart.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    chart.save(aggregate_mab_and_binding)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1409c7d-12b7-49d3-9b47-dda3ea6b7dbd",
   "metadata": {},
   "source": [
    "### Now show heatmap with nipah polymorphisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e84709f-3aef-4caf-8825-e7ac74a644aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_contact():\n",
    "    df = pd.DataFrame({\"site\": niv_poly, \"contact\": [0.0] * len(niv_poly)})\n",
    "    df = df[[\"site\", \"contact\"]]\n",
    "    # df['mutant'] = 'contact'\n",
    "    df[\"ab\"] = \"NiV Polymorphism\"\n",
    "    df[\"effect\"] = \"median_escape\"\n",
    "    df.rename(columns={\"contact\": \"value\"}, inplace=True)\n",
    "    return df\n",
    "\n",
    "\n",
    "niv_poly = [\n",
    "    82,\n",
    "    89,\n",
    "    135,\n",
    "    172,\n",
    "    228,\n",
    "    236,\n",
    "    274,\n",
    "    288,\n",
    "    299,\n",
    "    325,\n",
    "    328,\n",
    "    329,\n",
    "    335,\n",
    "    339,\n",
    "    344,\n",
    "    376,\n",
    "    384,\n",
    "    385,\n",
    "    386,\n",
    "    421,\n",
    "    423,\n",
    "    424,\n",
    "    426,\n",
    "    427,\n",
    "    470,\n",
    "    478,\n",
    "    481,\n",
    "    498,\n",
    "    502,\n",
    "    545,\n",
    "]\n",
    "contact_df = make_contact()\n",
    "\n",
    "bind_df = pd.read_csv(\"results/filtered_data/E2_binding_filtered.csv\")\n",
    "binding_df = bind_df.groupby(\"site\")[\"binding_median\"].max().reset_index()\n",
    "\n",
    "\n",
    "def make_empty_binding():\n",
    "    sites = range(71, 603)\n",
    "    data = [{\"site\": site} for site in sites]\n",
    "    empty_df = pd.DataFrame(data)\n",
    "    empty_df = pd.merge(empty_df, binding_df, on=\"site\", how=\"left\")\n",
    "    empty_df = empty_df.rename(columns={\"binding_median\": \"value\"})\n",
    "    empty_df[\"effect\"] = \"escape_median\"\n",
    "    empty_df[\"ab\"] = \"Ephrin-B2 binding\"\n",
    "    return empty_df\n",
    "\n",
    "\n",
    "binding_empty = make_empty_binding()\n",
    "\n",
    "escape_df = combined_df.groupby([\"ab\", \"site\"])[\"escape_median\"].max().reset_index()\n",
    "\n",
    "\n",
    "def make_empty_df(ab):\n",
    "    sites = range(71, 603)\n",
    "    data = [{\"site\": site} for site in sites]\n",
    "\n",
    "    # Create the DataFrame\n",
    "    empty_df = pd.DataFrame(data)\n",
    "\n",
    "    all_sites_df = pd.merge(\n",
    "        empty_df, escape_df.query(f'ab == \"{ab}\"'), on=[\"site\"], how=\"left\"\n",
    "    )\n",
    "\n",
    "    df_melted = all_sites_df.melt(\n",
    "        id_vars=[\"site\"],\n",
    "        value_vars=[\"escape_median\"],\n",
    "        var_name=\"effect\",\n",
    "        value_name=\"value\",\n",
    "    )\n",
    "\n",
    "    df_test = pd.concat([df_melted], ignore_index=True)\n",
    "    df_test[\"ab\"] = ab\n",
    "    return df_test\n",
    "\n",
    "\n",
    "ab_list = [\"m102.4\", \"HENV-26\", \"HENV-117\", \"HENV-103\", \"HENV-32\", \"nAH1.3\"]\n",
    "\n",
    "empty = []\n",
    "for ab in ab_list:\n",
    "    tmp_df = make_empty_df(ab)\n",
    "    empty.append(tmp_df)\n",
    "all_empties_df = pd.concat(empty, ignore_index=True)\n",
    "all_empties_df = pd.concat([all_empties_df, contact_df])\n",
    "display(all_empties_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f827d45-8693-4fac-962a-9771ed487e2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_heatmap_with_polymorphisms(df):\n",
    "    # Define the custom sort order directly in the encoding\n",
    "    sort_order = [\n",
    "        \"NiV Polymorphism\",\n",
    "        \"m102.4\",\n",
    "        \"HENV-26\",\n",
    "        \"HENV-117\",\n",
    "        \"HENV-103\",\n",
    "        \"HENV-32\",\n",
    "        \"nAH1.3\",\n",
    "    ]\n",
    "    # full_ranges = [list(range(start, end)) for start, end in [(71, 204), (204, 337), (337, 470), (470, 603)]]\n",
    "    full_ranges = [\n",
    "        list(range(start, end))\n",
    "        for start, end in [(71, 181), (181, 291), (291, 401), (401, 511), (511, 603)]\n",
    "    ]\n",
    "\n",
    "    # container to hold the charts\n",
    "    charts = []\n",
    "    color_scale_effect = alt.Scale(scheme=\"redblue\", domainMid=0, domain=[0, 2])\n",
    "    color_scale_binding = alt.Scale(scheme=\"redblue\", domainMid=0, domain=[-5, 2])\n",
    "\n",
    "    # Flags for showing the legend only the first time\n",
    "    effect_legend_added = True\n",
    "    binding_legend_added = True\n",
    "    for idx, subset in enumerate(full_ranges):\n",
    "        subset_df = df[df[\"site\"].isin(subset)]  # for the wrapping of sites\n",
    "        is_last_plot = idx == len(full_ranges) - 1\n",
    "        x_axis = alt.Axis(\n",
    "            labelAngle=-90,\n",
    "            labelExpr=\"datum.value % 10 === 0 ? datum.value : ''\",\n",
    "            title=\"Site\" if is_last_plot else None,\n",
    "            labels=True,\n",
    "        )  # Only show labels for the last plot\n",
    "\n",
    "        base = (\n",
    "            alt.Chart(subset_df)\n",
    "            .encode(\n",
    "                x=alt.X(\"site:O\", title=\"Site\", axis=x_axis),\n",
    "                y=alt.Y(\n",
    "                    \"ab\", title=None, sort=sort_order, axis=alt.Axis(grid=False)\n",
    "                ),  # Correctly apply custom sort order\n",
    "                tooltip=[\"site\", alt.Tooltip(\"value\", format=\".2f\")],\n",
    "            )\n",
    "            .properties(width=alt.Step(10), height=alt.Step(11))\n",
    "        )\n",
    "\n",
    "        # Define the chart for empty cells\n",
    "        chart_empty = base.mark_rect(color=\"#e6e7e8\").transform_filter(\n",
    "            alt.datum.effect == \"escape_median\"\n",
    "        )\n",
    "        if not effect_legend_added:\n",
    "            # Define the chart for cells with effect\n",
    "            chart_effect = (\n",
    "                base.mark_rect(stroke=\"black\", strokeWidth=0.25)\n",
    "                .encode(\n",
    "                    color=alt.condition(\n",
    "                        'datum.effect == \"escape_median\"',\n",
    "                        alt.Color(\n",
    "                            \"value:Q\", scale=color_scale_effect\n",
    "                        ),  # Define a color scale\n",
    "                        alt.value(\"transparent\"),\n",
    "                    )\n",
    "                )\n",
    "                .transform_filter(alt.datum.effect == \"escape_median\")\n",
    "            )\n",
    "            effect_legend_added = True\n",
    "        else:\n",
    "            # Define the chart for cells with effect\n",
    "            chart_effect = (\n",
    "                base.mark_rect(stroke=\"black\", strokeWidth=0.25)\n",
    "                .encode(\n",
    "                    color=alt.condition(\n",
    "                        'datum.effect == \"escape_median\"',\n",
    "                        alt.Color(\n",
    "                            \"value:Q\", scale=color_scale_effect, legend=None\n",
    "                        ),  # Define a color scale\n",
    "                        alt.value(\"transparent\"),\n",
    "                    )\n",
    "                )\n",
    "                .transform_filter(alt.datum.effect == \"escape_median\")\n",
    "            )\n",
    "        if not binding_legend_added:\n",
    "            chart_binding = (\n",
    "                base.mark_rect(strokeWidth=1.1)\n",
    "                .encode(\n",
    "                    stroke=alt.value(\"value\"),\n",
    "                    color=alt.condition(\n",
    "                        'datum.effect == \"escape_median\"',\n",
    "                        alt.Color(\"value:Q\", scale=color_scale_binding),\n",
    "                        alt.value(\"transparent\"),\n",
    "                    ),\n",
    "                )\n",
    "                .transform_filter(alt.datum.ab == \"Ephrin-B2 binding\")\n",
    "            )\n",
    "            binding_legend_added = True\n",
    "        else:\n",
    "            chart_binding = (\n",
    "                base.mark_rect(strokeWidth=1.1)\n",
    "                .encode(\n",
    "                    stroke=alt.value(\"value\"),\n",
    "                    color=alt.condition(\n",
    "                        'datum.effect == \"escape_median\"',\n",
    "                        alt.Color(\"value:Q\", scale=color_scale_binding, legend=None),\n",
    "                        alt.value(\"transparent\"),\n",
    "                    ),\n",
    "                )\n",
    "                .transform_filter(alt.datum.ab == \"Ephrin-B2 binding\")\n",
    "            )\n",
    "\n",
    "        chart_poly = (\n",
    "            base.mark_rect(color=\"black\")\n",
    "            .encode()\n",
    "            .transform_filter(alt.datum.ab == \"NiV Polymorphism\")\n",
    "        )\n",
    "        # Layer the charts using `layer` instead of `+`\n",
    "        chart = alt.layer(chart_empty, chart_effect, chart_poly).resolve_scale(\n",
    "            color=\"independent\"\n",
    "        )\n",
    "        charts.append(chart)\n",
    "    combined_chart = alt.vconcat(\n",
    "        *charts, spacing=5, title=\"Heatmap of max mAb escape and Nipah Polymorphisms\"\n",
    "    ).resolve_scale(y=\"shared\", x=\"independent\", color=\"shared\")\n",
    "\n",
    "    return combined_chart\n",
    "\n",
    "\n",
    "# Assuming `all_empties_df` is your DataFrame and already defined\n",
    "chart = make_heatmap_with_polymorphisms(all_empties_df)\n",
    "chart.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    chart.save(aggregate_mab_and_niv_polymorphism)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b440e7f9-ebe2-4460-8025-9b7d3b909672",
   "metadata": {},
   "source": [
    "### Make plots comparing escape with binding to see if escape sites do so by increasing binding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4553d7a-56fa-425a-92fa-f9a0ea042212",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "new_merged_df = pd.merge(\n",
    "    combined_df,\n",
    "    bind_df[[\"site\", \"wildtype\", \"mutant\", \"binding_median\"]],\n",
    "    on=[\"site\", \"wildtype\", \"mutant\"],\n",
    "    how=\"left\",\n",
    ")\n",
    "new_merged_df = new_merged_df.drop(\n",
    "    columns=[\n",
    "        \"mutation\",\n",
    "        \"escape_std\",\n",
    "        \"times_seen_ab\",\n",
    "        \"show_site\",\n",
    "        \"wt_codon\",\n",
    "        \"closest_mutant_codon\",\n",
    "        \"min_mutations\",\n",
    "    ]\n",
    ")\n",
    "new_merged_df = new_merged_df.round(2)\n",
    "\n",
    "ab_list1 = [\"m102.4\", \"HENV-26\", \"HENV-117\"]\n",
    "ab_list2 = [\"HENV-103\", \"HENV-32\"]\n",
    "ab_list3 = [\"nAH1.3\"]\n",
    "\n",
    "\n",
    "def plot_escape_vs_binding(df):\n",
    "    variant_selector = alt.selection_point(\n",
    "        on=\"mouseover\", empty=False, nearest=True, fields=[\"site\"], value=1\n",
    "    )\n",
    "    empty_chart1 = []\n",
    "    for ab in ab_list1:\n",
    "        tmp_df = df[df[\"ab\"] == ab]\n",
    "        base = (\n",
    "            alt.Chart(tmp_df, title=alt.Title(f\"{ab}\", anchor=\"middle\"))\n",
    "            .mark_point(\n",
    "                filled=True, size=15, color=\"#1f4e79\", opacity=0.15, stroke=\"black\"\n",
    "            )\n",
    "            .encode(\n",
    "                alt.X(\n",
    "                    \"binding_median\",\n",
    "                    title=\"EFNB2 Binding\",\n",
    "                    axis=alt.Axis(grid=True, tickCount=3),\n",
    "                ),\n",
    "                alt.Y(\n",
    "                    \"escape_median\",\n",
    "                    title=\"Antibody Escape\",\n",
    "                    axis=alt.Axis(grid=True, tickCount=3),\n",
    "                ),\n",
    "                tooltip=[\n",
    "                    \"site\",\n",
    "                    \"wildtype\",\n",
    "                    \"mutant\",\n",
    "                    \"escape_median\",\n",
    "                    \"binding_median\",\n",
    "                ],\n",
    "                opacity=alt.condition(variant_selector, alt.value(1), alt.value(0.2)),\n",
    "                size=alt.condition(variant_selector, alt.value(50), alt.value(20)),\n",
    "                strokeWidth=alt.condition(variant_selector, alt.value(2), alt.value(0)),\n",
    "            )\n",
    "        )\n",
    "        empty_chart1.append(base)\n",
    "    combined_chart1 = alt.hconcat(*empty_chart1, spacing=5).resolve_scale(\n",
    "        x=\"shared\", y=\"shared\"\n",
    "    )\n",
    "    empty_chart2 = []\n",
    "    for ab in ab_list2:\n",
    "        tmp_df = df[df[\"ab\"] == ab]\n",
    "        base = (\n",
    "            alt.Chart(tmp_df, title=alt.Title(f\"{ab}\", anchor=\"middle\"))\n",
    "            .mark_point(\n",
    "                filled=True, size=15, color=\"#ff7f0e\", opacity=0.15, stroke=\"black\"\n",
    "            )\n",
    "            .encode(\n",
    "                alt.X(\n",
    "                    \"binding_median\",\n",
    "                    title=\"EFNB2 Binding\",\n",
    "                    axis=alt.Axis(grid=True, tickCount=3),\n",
    "                ),\n",
    "                alt.Y(\n",
    "                    \"escape_median\",\n",
    "                    title=\"Antibody Escape\",\n",
    "                    axis=alt.Axis(grid=True, tickCount=3),\n",
    "                ),\n",
    "                tooltip=[\n",
    "                    \"site\",\n",
    "                    \"wildtype\",\n",
    "                    \"mutant\",\n",
    "                    \"escape_median\",\n",
    "                    \"binding_median\",\n",
    "                ],\n",
    "                opacity=alt.condition(variant_selector, alt.value(1), alt.value(0.2)),\n",
    "                size=alt.condition(variant_selector, alt.value(50), alt.value(20)),\n",
    "                strokeWidth=alt.condition(variant_selector, alt.value(2), alt.value(0)),\n",
    "            )\n",
    "        )\n",
    "        empty_chart2.append(base)\n",
    "    combined_chart2 = alt.hconcat(*empty_chart2, spacing=5).resolve_scale(\n",
    "        x=\"shared\", y=\"shared\"\n",
    "    )\n",
    "\n",
    "    empty_chart3 = []\n",
    "    for ab in ab_list3:\n",
    "        tmp_df = df[df[\"ab\"] == ab]\n",
    "        base3 = (\n",
    "            alt.Chart(tmp_df, title=alt.Title(f\"{ab}\", anchor=\"middle\"))\n",
    "            .mark_point(\n",
    "                filled=True, size=15, color=\"#2ca02c\", opacity=0.15, stroke=\"black\"\n",
    "            )\n",
    "            .encode(\n",
    "                alt.X(\n",
    "                    \"binding_median\",\n",
    "                    title=\"EFNB2 Binding\",\n",
    "                    axis=alt.Axis(grid=True, tickCount=3),\n",
    "                ),\n",
    "                alt.Y(\n",
    "                    \"escape_median\",\n",
    "                    title=\"Antibody Escape\",\n",
    "                    axis=alt.Axis(grid=True, tickCount=3),\n",
    "                ),\n",
    "                tooltip=[\n",
    "                    \"site\",\n",
    "                    \"wildtype\",\n",
    "                    \"mutant\",\n",
    "                    \"escape_median\",\n",
    "                    \"binding_median\",\n",
    "                ],\n",
    "                opacity=alt.condition(variant_selector, alt.value(1), alt.value(0.2)),\n",
    "                size=alt.condition(variant_selector, alt.value(50), alt.value(20)),\n",
    "                strokeWidth=alt.condition(variant_selector, alt.value(2), alt.value(0)),\n",
    "            )\n",
    "        )\n",
    "\n",
    "    combined_chart_total = alt.vconcat(\n",
    "        combined_chart1,\n",
    "        combined_chart2,\n",
    "        base3,\n",
    "        title=alt.Title(\n",
    "            \"Antibody Escape versus Binding\",\n",
    "            subtitle=\"Colored by Epitope. Hover over points to see the same sites\",\n",
    "        ),\n",
    "    ).add_params(\n",
    "        variant_selector\n",
    "    )  \n",
    "    return combined_chart_total\n",
    "\n",
    "\n",
    "tmp_img_test = plot_escape_vs_binding(new_merged_df)\n",
    "tmp_img_test.display()\n",
    "if mab_line_escape_plot is not None:\n",
    "    tmp_img_test.save(binding_vs_escape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40788f82-688c-4eaf-82e0-f16d9880b4ac",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
