<!DOCTYPE html>
<html lang="en-US" dir="ltr">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Pipeline Information | Nipah RBP DMS</title>
    <meta name="description" content="Information about Nipah RBP DMS experiments">
    <meta name="generator" content="VitePress v1.0.2">
    <link rel="preload stylesheet" href="/Nipah_Malaysia_RBP_DMS/assets/style.DK17GJ8M.css" as="style">
    
    <script type="module" src="/Nipah_Malaysia_RBP_DMS/assets/app.CvytJhZQ.js"></script>
    <link rel="preload" href="/Nipah_Malaysia_RBP_DMS/assets/inter-roman-latin.Bu8hRsVA.woff2" as="font" type="font/woff2" crossorigin="">
    <link rel="modulepreload" href="/Nipah_Malaysia_RBP_DMS/assets/chunks/framework.CkjTfQJn.js">
    <link rel="modulepreload" href="/Nipah_Malaysia_RBP_DMS/assets/chunks/theme.Ds4xM_o9.js">
    <link rel="modulepreload" href="/Nipah_Malaysia_RBP_DMS/assets/pipeline_information.md.BjJCubcu.lean.js">
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-K1YW76KQ9L"></script>
    <script>window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments)}gtag("js",new Date),gtag("config","G-K1YW76KQ9L");</script>
    <script id="check-dark-mode">(()=>{const e=localStorage.getItem("vitepress-theme-appearance")||"auto",a=window.matchMedia("(prefers-color-scheme: dark)").matches;(!e||e==="auto"?a:e==="dark")&&document.documentElement.classList.add("dark")})();</script>
    <script id="check-mac-os">document.documentElement.classList.toggle("mac",/Mac|iPhone|iPod|iPad/i.test(navigator.platform));</script>
  </head>
  <body>
    <div id="app"><div class="Layout" data-v-5d98c3a5><!--[--><!--]--><!--[--><span tabindex="-1" data-v-0f60ec36></span><a href="#VPContent" class="VPSkipLink visually-hidden" data-v-0f60ec36> Skip to content </a><!--]--><!----><header class="VPNav" data-v-5d98c3a5 data-v-ae24b3ad><div class="VPNavBar top" data-v-ae24b3ad data-v-ccf7ddec><div class="wrapper" data-v-ccf7ddec><div class="container" data-v-ccf7ddec><div class="title" data-v-ccf7ddec><div class="VPNavBarTitle" data-v-ccf7ddec data-v-ab179fa1><a class="title" href="/Nipah_Malaysia_RBP_DMS/" data-v-ab179fa1><!--[--><!--]--><!----><span data-v-ab179fa1>Nipah RBP DMS</span><!--[--><!--]--></a></div></div><div class="content" data-v-ccf7ddec><div class="content-body" data-v-ccf7ddec><!--[--><!--]--><div class="VPNavBarSearch search" data-v-ccf7ddec><!----></div><nav aria-labelledby="main-nav-aria-label" class="VPNavBarMenu menu" data-v-ccf7ddec data-v-7f418b0f><span id="main-nav-aria-label" class="visually-hidden" data-v-7f418b0f>Main Navigation</span><!--[--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/Nipah_Malaysia_RBP_DMS/" tabindex="0" data-v-7f418b0f data-v-9c663999><!--[--><span data-v-9c663999>Home</span><!--]--></a><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink active" href="/Nipah_Malaysia_RBP_DMS/pipeline_information.html" tabindex="0" data-v-7f418b0f data-v-9c663999><!--[--><span data-v-9c663999>Analysis Pipeline</span><!--]--></a><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/Nipah_Malaysia_RBP_DMS/cell_entry.html" tabindex="0" data-v-7f418b0f data-v-9c663999><!--[--><span data-v-9c663999>Cell Entry</span><!--]--></a><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/Nipah_Malaysia_RBP_DMS/receptor_binding.html" tabindex="0" data-v-7f418b0f data-v-9c663999><!--[--><span data-v-9c663999>Receptor Binding</span><!--]--></a><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/Nipah_Malaysia_RBP_DMS/antibody_escape.html" tabindex="0" data-v-7f418b0f data-v-9c663999><!--[--><span data-v-9c663999>Antibody Escape</span><!--]--></a><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/Nipah_Malaysia_RBP_DMS/heatmaps.html" tabindex="0" data-v-7f418b0f data-v-9c663999><!--[--><span data-v-9c663999>Heatmaps</span><!--]--></a><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/Nipah_Malaysia_RBP_DMS/interactive.html" tabindex="0" data-v-7f418b0f data-v-9c663999><!--[--><span data-v-9c663999>Interactive Figures</span><!--]--></a><!--]--><!--]--></nav><!----><div class="VPNavBarAppearance appearance" data-v-ccf7ddec data-v-e6aabb21><button class="VPSwitch VPSwitchAppearance" type="button" role="switch" title="Switch to dark theme" aria-checked="false" data-v-e6aabb21 data-v-d1f28634 data-v-1d5665e3><span class="check" data-v-1d5665e3><span class="icon" data-v-1d5665e3><!--[--><span class="vpi-sun sun" data-v-d1f28634></span><span class="vpi-moon moon" data-v-d1f28634></span><!--]--></span></span></button></div><div class="VPSocialLinks VPNavBarSocialLinks social-links" data-v-ccf7ddec data-v-0394ad82 data-v-7bc22406><!--[--><a class="VPSocialLink no-icon" href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS" aria-label="github" target="_blank" rel="noopener" data-v-7bc22406 data-v-eee4e7cb><span class="vpi-social-github" /></a><!--]--></div><div class="VPFlyout VPNavBarExtra extra" data-v-ccf7ddec data-v-d0bd9dde data-v-b6c34ac9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" aria-label="extra navigation" data-v-b6c34ac9><span class="vpi-more-horizontal icon" data-v-b6c34ac9></span></button><div class="menu" data-v-b6c34ac9><div class="VPMenu" data-v-b6c34ac9 data-v-e7ea1737><!----><!--[--><!--[--><!----><div class="group" data-v-d0bd9dde><div class="item appearance" data-v-d0bd9dde><p class="label" data-v-d0bd9dde>Appearance</p><div class="appearance-action" data-v-d0bd9dde><button class="VPSwitch VPSwitchAppearance" type="button" role="switch" title="Switch to dark theme" aria-checked="false" data-v-d0bd9dde data-v-d1f28634 data-v-1d5665e3><span class="check" data-v-1d5665e3><span class="icon" data-v-1d5665e3><!--[--><span class="vpi-sun sun" data-v-d1f28634></span><span class="vpi-moon moon" data-v-d1f28634></span><!--]--></span></span></button></div></div></div><div class="group" data-v-d0bd9dde><div class="item social-links" data-v-d0bd9dde><div class="VPSocialLinks social-links-list" data-v-d0bd9dde data-v-7bc22406><!--[--><a class="VPSocialLink no-icon" href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS" aria-label="github" target="_blank" rel="noopener" data-v-7bc22406 data-v-eee4e7cb><span class="vpi-social-github" /></a><!--]--></div></div></div><!--]--><!--]--></div></div></div><!--[--><!--]--><button type="button" class="VPNavBarHamburger hamburger" aria-label="mobile navigation" aria-expanded="false" aria-controls="VPNavScreen" data-v-ccf7ddec data-v-e5dd9c1c><span class="container" data-v-e5dd9c1c><span class="top" data-v-e5dd9c1c></span><span class="middle" data-v-e5dd9c1c></span><span class="bottom" data-v-e5dd9c1c></span></span></button></div></div></div></div><div class="divider" data-v-ccf7ddec><div class="divider-line" data-v-ccf7ddec></div></div></div><!----></header><div class="VPLocalNav empty fixed" data-v-5d98c3a5 data-v-a6f0e41e><div class="container" data-v-a6f0e41e><!----><div class="VPLocalNavOutlineDropdown" style="--vp-vh:0px;" data-v-a6f0e41e data-v-d2ecc192><button data-v-d2ecc192>Return to top</button><!----></div></div></div><!----><div class="VPContent" id="VPContent" data-v-5d98c3a5 data-v-1428d186><div class="VPDoc has-aside" data-v-1428d186 data-v-39a288b8><!--[--><!--]--><div class="container" data-v-39a288b8><div class="aside" data-v-39a288b8><div class="aside-curtain" data-v-39a288b8></div><div class="aside-container" data-v-39a288b8><div class="aside-content" data-v-39a288b8><div class="VPDocAside" data-v-39a288b8 data-v-3f215769><!--[--><!--]--><!--[--><!--]--><div class="VPDocAsideOutline" role="navigation" data-v-3f215769 data-v-935f8a84><div class="content" data-v-935f8a84><div class="outline-marker" data-v-935f8a84></div><div class="outline-title" role="heading" aria-level="2" data-v-935f8a84>On this page</div><nav aria-labelledby="doc-outline-aria-label" data-v-935f8a84><span class="visually-hidden" id="doc-outline-aria-label" data-v-935f8a84> Table of Contents for current page </span><ul class="VPDocOutlineItem root" data-v-935f8a84 data-v-b933a997><!--[--><!--]--></ul></nav></div></div><!--[--><!--]--><div class="spacer" data-v-3f215769></div><!--[--><!--]--><!----><!--[--><!--]--><!--[--><!--]--></div></div></div></div><div class="content" data-v-39a288b8><div class="content-container" data-v-39a288b8><!--[--><!--]--><main class="main" data-v-39a288b8><div style="position:relative;" class="vp-doc _Nipah_Malaysia_RBP_DMS_pipeline_information" data-v-39a288b8><div><h1 id="pipeline-information" tabindex="-1">Pipeline Information <a class="header-anchor" href="#pipeline-information" aria-label="Permalink to &quot;Pipeline Information&quot;">â€‹</a></h1><p>The lentivirus-based deep mutational scanning platform relies on obtaining relative frequencies of different receptor binding protein variants that enter cells after applying selection to the libraries. By comparing mutation frequencies before and after selections, we can determine the effects of mutations on different phenotypes.</p><p>Calculating the relative frequencies of thousands of variants is not trivial. We rely on two different sequencing technologies to obtain the necessary data.</p><ul><li>PacBio long-read sequencing to link barcodes to specific mutations in the receptor binding protein.</li><li>Illumina short-read sequencing to obtain the relative frequencies of barcodes in each selection experiment.</li></ul><p><img src="/Nipah_Malaysia_RBP_DMS/assets/library_schematic.BON00GhX.png" alt="image"></p><div style="text-align:center;color:grey;font-size:smaller;">Schematic of lentivirus vector used in deep mutational scanning experiments (top), along with sequencing strategy (bottom). </div><p>Because PacBio sequencing is expensive and lower throughput, we only sequenced the variant libraries with this technology once. Full-length consensus sequences of the receptor binding protein and associated barcodes are assembled, while discarding low-quality reads. From these assembled consensus sequences we build a codon-variant lookup table, enabling us to match barcodes to specific mutations in the receptor binding protein. All subsequent Illumina sequencing of selection experiments use this lookup table to estimate mutational effects from barcode sequencing data alone. Our generated pseudovirus libraries consist of 60,000 to 80,000 unique variants. Each unique variant is sequenced hundreds of times with Illumina to get accurate frequency measurements.</p><p>Most of these computationally intensive steps were analyzed with <a href="https://github.com/dms-vep/dms-vep-pipeline-3" target="_blank" rel="noreferrer"><code>dms-vep-pipeline-3</code></a>. This pipeline utilizes the <a href="https://jbloomlab.github.io/alignparse/" target="_blank" rel="noreferrer"><code>alignparse</code></a> package. Each step, along with the associated jupyter notebooks are listed below.</p><h2 id="build-pacbio-sequences" tabindex="-1">Build Pacbio Sequences <a class="header-anchor" href="#build-pacbio-sequences" aria-label="Permalink to &quot;Build Pacbio Sequences&quot;">â€‹</a></h2><p><a href="notebooks/build_pacbio_consensus.html" target="_self">PacBio consensus sequences notebook</a></p><p>This notebook builds the Pacbio consensus sequences used to link specific mutations in the receptor binding protein with a unique 16 bp barcode. <a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/config.yaml" target="_blank" rel="noreferrer">Parameters</a> used are:</p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code"><code><span class="line"><span>max_minor_sub_frac=0.2</span></span>
<span class="line"><span>max_minor_indel_frac=0.2</span></span>
<span class="line"><span>min_support=3</span></span></code></pre></div><p>These parameters filter consensus sequences generated from Pacbio CCS sequencing and assembly. If an assembled RBP sequence has a mutation or indel in more than 20% of the reads, it will be discarded. Consensus sequences must have at least three reads to be included as variants.</p><p>Using <a href="https://jbloomlab.github.io/alignparse/" target="_blank" rel="noreferrer">alignparse</a>, reads were mapped to a <a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/data/PacBio_amplicon.gb" target="_blank" rel="noreferrer">reference sequence</a>, and clipped based on parameters in this <a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/data/PacBio_feature_parse_specs.yaml" target="_blank" rel="noreferrer">config</a> file.</p><h2 id="analyze-pacbio-ccs-reads" tabindex="-1">Analyze PacBio CCS Reads <a class="header-anchor" href="#analyze-pacbio-ccs-reads" aria-label="Permalink to &quot;Analyze PacBio CCS Reads&quot;">â€‹</a></h2><p><a href="notebooks/analyze_pacbio_ccs.html" target="_self">Analyze Pacbio CCS reads notebook</a></p><p>Reports information about CCS read filtering.</p><h2 id="build-codon-variants-notebook" tabindex="-1">Build Codon Variants Notebook <a class="header-anchor" href="#build-codon-variants-notebook" aria-label="Permalink to &quot;Build Codon Variants Notebook&quot;">â€‹</a></h2><p><a href="notebooks/build_codon_variants.html" target="_self">Build codon variants notebook</a></p><p>Builds the codon-variant table from PacBio consensus sequences that links barcodes and RBP mutations. Displays information about the number of mutations and variants present in each library.</p><p><a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/results/variants/codon_variants.csv" target="_blank" rel="noreferrer">Link to codon-variant table .csv file</a></p><h2 id="illumina-variant-counts" tabindex="-1">Illumina Variant Counts <a class="header-anchor" href="#illumina-variant-counts" aria-label="Permalink to &quot;Illumina Variant Counts&quot;">â€‹</a></h2><p>Once the barcodes are linked to mutations in the codon-variant table, all sequencing data is generated with Illumina on a small sequence fragment to obtain the relative frequencies of barcodes in each selection experiment. The config file linked below specifies the parameters used for converting barcode counts to functional scores, which are used to estimate cell entry.</p><p><a href="notebooks/analyze_variant_counts.html" target="_self">Analysis of variant counts notebook</a></p><p><a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/tree/master/results/barcode_counts" target="_blank" rel="noreferrer">Link to raw barcode count .csv files</a></p><p><a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/data/func_effects_config.yml" target="_blank" rel="noreferrer">Link to functional selection config file</a></p><h2 id="filtering-selection-data" tabindex="-1">Filtering Selection Data <a class="header-anchor" href="#filtering-selection-data" aria-label="Permalink to &quot;Filtering Selection Data&quot;">â€‹</a></h2><p><a href="notebooks/filter_data.html" target="_self">Filtering notebook</a></p><p>Once the effects of mutations on different phenotypes have been calculated, we perform a data filtering step to remove low confidence measurements. The filtering parameters are contained within the <a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/nipah_config.yaml" target="_blank" rel="noreferrer">nipah_config.yaml</a> file. More information about these parameters are listed in the notebook. In brief, we require mutations to be present in at least two barcodes, and have low variance between selection replicates.</p><h2 id="filtered-data" tabindex="-1">Filtered Data <a class="header-anchor" href="#filtered-data" aria-label="Permalink to &quot;Filtered Data&quot;">â€‹</a></h2><div class="warning custom-block"><p class="custom-block-title">These data have been filtered and are the best choice for anyone interested in analyzing the data themselves. For unfiltered raw <code>.csv</code> files of mutational effects on different phenotypes, go to individual pages to view and download.</p><h3 id="cell-entry" tabindex="-1">Cell Entry <a class="header-anchor" href="#cell-entry" aria-label="Permalink to &quot;Cell Entry&quot;">â€‹</a></h3><p><a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/results/filtered_data/public_filtered/RBP_mutation_effects_cell_entry_CHO-bEFNB2.csv" target="_blank" rel="noreferrer">CHO-bEFNB2 entry filtered (.csv)</a></p><p><a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/results/filtered_data/public_filtered/RBP_mutation_effects_cell_entry_CHO-bEFNB3.csv" target="_blank" rel="noreferrer">CHO-bEFNB3 entry filtered (.csv)</a></p><h3 id="receptor-binding" tabindex="-1">Receptor Binding <a class="header-anchor" href="#receptor-binding" aria-label="Permalink to &quot;Receptor Binding&quot;">â€‹</a></h3><p><a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/results/filtered_data/public_filtered/RBP_mutation_effects_bEFNB2_binding.csv" target="_blank" rel="noreferrer">bEFNB2 monomeric binding filtered (.csv)</a></p><p><a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/results/filtered_data/public_filtered/RBP_mutation_effects_bEFNB3_binding.csv" target="_blank" rel="noreferrer">bEFNB3 dimeric binding filtered (.csv)</a></p><h3 id="antibody-escape" tabindex="-1">Antibody Escape <a class="header-anchor" href="#antibody-escape" aria-label="Permalink to &quot;Antibody Escape&quot;">â€‹</a></h3><p><a href="https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS/blob/master/results/filtered_data/public_filtered/RBP_mutation_effects_antibody_escape.csv" target="_blank" rel="noreferrer">Antibody escape filtered (.csv)</a></p></div><h2 id="miscellaneous-notebooks" tabindex="-1">Miscellaneous Notebooks <a class="header-anchor" href="#miscellaneous-notebooks" aria-label="Permalink to &quot;Miscellaneous Notebooks&quot;">â€‹</a></h2><p><a href="notebooks/library_correlations.html" target="_self">Notebook for finding correlations between libraries and making histogram of variants</a></p><p><a href="notebooks/make_nipah_phylogeny_baltic.html" target="_self">Notebook for making a Nipah phylogeny</a></p><p><a href="notebooks/mapping_site_level.html" target="_self">Notebook for making specific file formats (.defattr), to map site-averaged scores onto protein structures in Chimera</a></p><p><a href="notebooks/receptor_distance.html" target="_self">Notebook for calculating atomic distances between residues from a PDB file</a></p><p><a href="notebooks/henipavirus_conservation.html" target="_self">Notebook for finding variable sites in Nipah or Henipavirus alignments</a></p></div></div></main><footer class="VPDocFooter" data-v-39a288b8 data-v-09de1c0f><!--[--><!--]--><!----><!----></footer><!--[--><!--]--></div></div></div><!--[--><!--]--></div></div><footer class="VPFooter" data-v-5d98c3a5 data-v-e315a0ad><div class="container" data-v-e315a0ad><p class="message" data-v-e315a0ad>Built by Brendan Larsen and Jesse Bloom</p><!----></div></footer><!--[--><!--]--></div></div>
    <script>window.__VP_HASH_MAP__=JSON.parse("{\"index.md\":\"Bx2skdmA\",\"interactive.md\":\"BrWcZFR1\",\"pipeline_information.md\":\"BjJCubcu\",\"cell_entry.md\":\"BCpXmNnB\",\"receptor_binding.md\":\"GTSRVDbQ\",\"antibody_escape.md\":\"DWTMqAE3\",\"heatmaps.md\":\"C4mrizPX\"}");window.__VP_SITE_DATA__=JSON.parse("{\"lang\":\"en-US\",\"dir\":\"ltr\",\"title\":\"Nipah RBP DMS\",\"description\":\"Information about Nipah RBP DMS experiments\",\"base\":\"/Nipah_Malaysia_RBP_DMS/\",\"head\":[],\"router\":{\"prefetchLinks\":true},\"appearance\":true,\"themeConfig\":{\"nav\":[{\"text\":\"Home\",\"link\":\"/\"},{\"text\":\"Analysis Pipeline\",\"link\":\"/pipeline_information\"},{\"text\":\"Cell Entry\",\"link\":\"/cell_entry\"},{\"text\":\"Receptor Binding\",\"link\":\"/receptor_binding\"},{\"text\":\"Antibody Escape\",\"link\":\"/antibody_escape\"},{\"text\":\"Heatmaps\",\"link\":\"/heatmaps\"},{\"text\":\"Interactive Figures\",\"link\":\"/interactive\"}],\"socialLinks\":[{\"icon\":\"github\",\"link\":\"https://github.com/dms-vep/Nipah_Malaysia_RBP_DMS\"}],\"footer\":{\"message\":\"Built by Brendan Larsen and Jesse Bloom\"}},\"locales\":{},\"scrollOffset\":134,\"cleanUrls\":false}");</script>
    
  </body>
</html>